<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Annotating the human IGH locus &mdash; Digger 0.5.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Annotating the rhesus macaque IGH locus" href="rhesus_igh.html" />
    <link rel="prev" title="Release Notes" href="../news.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Digger
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../overview.html#digger"><code class="docutils literal notranslate"><span class="pre">digger</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../overview.html#dig-sequence"><code class="docutils literal notranslate"><span class="pre">dig-sequence</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../news.html">Release Notes</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Annotating the human IGH locus</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#data">Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#annotating-the-assembly">Annotating the Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="#comparing-the-output-to-imgt-s-annotation">Comparing the output to IMGT’s annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="rhesus_igh.html">Annotating the rhesus macaque IGH locus</a></li>
<li class="toctree-l1"><a class="reference internal" href="targeted_annotation.html">Targeted Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="additional_examples.html">Additional Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Commandline Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools/annotation.html">Anotation format</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Digger</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Annotating the human IGH locus</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/examples/human_igh.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="annotating-the-human-igh-locus">
<span id="human-igh"></span><h1>Annotating the human IGH locus<a class="headerlink" href="#annotating-the-human-igh-locus" title="Permalink to this heading"></a></h1>
<p>The IGHV locus in the human reference assembly GRCh38.p12 has been annotated by IMGT. In this example, we will annotate with Digger and compare results.
The comparison, and a script to reproduce it using the steps below, can be found <a class="reference external" href="https://github.com/williamdlees/digger/tree/main/tests/human/IGH/IMGT000035">in digger’s GitHub repository</a>.
Digger’s annotation <a class="reference external" href="https://github.com/williamdlees/digger/blob/main/tests/human/IGH/IMGT000035/IMGT000035.csv">can also be found</a> in that directory.</p>
<section id="data">
<h2>Data<a class="headerlink" href="#data" title="Permalink to this heading"></a></h2>
<p>The locus and IMGT annotations can be downloaded and parsed with <a class="reference internal" href="../tools/parse_imgt_annotations.html#parse-imgt-annotations"><span class="std std-ref">parse_imgt_annotations</span></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">parse_imgt_annotations</span> \
    <span class="o">--</span><span class="n">save_sequence</span> <span class="n">IMGT000035</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="s2">&quot;http://www.imgt.org/ligmdb/view?format=IMGT&amp;id=IMGT000035&quot;</span> \
        <span class="n">IMGT000035_genes</span><span class="o">.</span><span class="n">csv</span> \
        <span class="n">IGH</span>
</pre></div>
</div>
<p>This will create two files: <code class="docutils literal notranslate"><span class="pre">IMGT000035.fasta</span></code>, containing the assembly sequence, and <code class="docutils literal notranslate"><span class="pre">IMGT000035_genes.csv</span></code>, containing the co-ordinates of the IMGT-annotated genes, and sequences of their flanking regions.</p>
<p>The human IGH reference set can be downloaded from IMGT with the <a class="reference external" href="https://williamdlees.github.io/receptor_utils/_build/html/introduction.html">receptor_utils</a> command <code class="docutils literal notranslate"><span class="pre">extract_refs</span></code> (receptor_utils is installed as part of Digger’s installation):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">extract_refs</span> <span class="o">-</span><span class="n">L</span> <span class="n">IGH</span> <span class="s2">&quot;Homo sapiens&quot;</span>
</pre></div>
</div>
<p>This creates the reference sets <code class="docutils literal notranslate"><span class="pre">Homo_sapiens_IGHV_gapped.fasta</span></code>, <code class="docutils literal notranslate"><span class="pre">Homo_sapiens_IGHV.fasta</span></code>, <code class="docutils literal notranslate"><span class="pre">Homo_sapiens_IGHD.fasta</span></code>, <code class="docutils literal notranslate"><span class="pre">Homo_sapiens_IGHJ.fasta</span></code>, <code class="docutils literal notranslate"><span class="pre">Homo_sapiens_CH.fasta</span></code></p>
<p>Lastly, we need to make a combined file containing all the IGH V, D and J reference genes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">cat</span> <span class="n">Homo_sapiens_IGHV</span><span class="o">.</span><span class="n">fasta</span> <span class="n">Homo_sapiens_IGHD</span><span class="o">.</span><span class="n">fasta</span> <span class="n">Homo_sapiens_IGHJ</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">&gt;</span> <span class="n">Homo_sapiens_IGHVDJ</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
</section>
<section id="annotating-the-assembly">
<h2>Annotating the Assembly<a class="headerlink" href="#annotating-the-assembly" title="Permalink to this heading"></a></h2>
<p>With the data in place, we can instruct <a class="reference internal" href="../tools/digger.html#digger"><span class="std std-ref">digger</span></a> to perform the annotation:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">digger</span> <span class="n">IMGT000035</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">-</span><span class="n">v_ref</span> <span class="n">Homo_sapiens_IGHV</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">-</span><span class="n">d_ref</span> <span class="n">Homo_sapiens_IGHD</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">-</span><span class="n">j_ref</span> <span class="n">Homo_sapiens_IGHJ</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">-</span><span class="n">v_ref_gapped</span> <span class="n">Homo_sapiens_IGHV_gapped</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">-</span><span class="n">ref</span> <span class="n">imgt</span><span class="p">,</span><span class="n">Homo_sapiens_IGHVDJ</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">-</span><span class="n">species</span> <span class="n">human</span> \
        <span class="o">-</span><span class="n">locus</span> <span class="n">IGH</span> \
        <span class="n">IMGT000035</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">-v_ref,</span> <span class="pre">-d_ref,</span> <span class="pre">-j_ref,</span> <span class="pre">-v_ref_gapped</span></code> provide the germline reference sequences.</p>
<p><code class="docutils literal notranslate"><span class="pre">-ref</span> <span class="pre">imgt,Homo_sapiens_IGHVDJ.fasta</span></code> instructs Digger to compare any sequences it identifies in the assembly with those in the combined reference
set. It will create columns in the output assembly containing the nearest reference sequence found, % identity and so on. These columns will be prefixed <code class="docutils literal notranslate"><span class="pre">imgt</span></code> as specified in this argument. Multiple <code class="docutils literal notranslate"><span class="pre">-ref</span></code> arguments can be
passed, allowing comparison with multiple reference sets.</p>
<p><code class="docutils literal notranslate"><span class="pre">-species</span></code> tells Digger to use its internal position-weighted matrices for human loci, and <code class="docutils literal notranslate"><span class="pre">-locus</span></code> specifies the locus.</p>
<p>The output file will be <code class="docutils literal notranslate"><span class="pre">IMGT000035.csv</span></code>. A sample copy of this file will be found in the GitHub repository.</p>
<p>Digger will summarise progress as it runs. It will call the following tools:</p>
<ul class="simple">
<li><p>makeblastdb to create BLAST databases for the reference genes</p></li>
<li><p>blastn to run these databases against the reference sequence</p></li>
<li><p><a class="reference internal" href="../tools/blastresults_to_csv.html#blastresults-to-csv"><span class="std std-ref">blastresults_to_csv</span></a> to convert the BLAST output to a simpler format</p></li>
<li><p><a class="reference internal" href="../tools/find_alignments.html#find-alignments"><span class="std std-ref">find_alignments</span></a> to annotate gene alignments found in the assembly</p></li>
</ul>
</section>
<section id="comparing-the-output-to-imgt-s-annotation">
<h2>Comparing the output to IMGT’s annotation<a class="headerlink" href="#comparing-the-output-to-imgt-s-annotation" title="Permalink to this heading"></a></h2>
<p><a class="reference internal" href="../tools/compare_annotations.html#compare-annotations"><span class="std std-ref">compare_annotations</span></a> compares the output of a digger run with a summary annotation file. Here we use it to compare the results of digger’s annotation with IMGT’s:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">compare_annotations</span> <span class="n">IMGT000035</span><span class="o">.</span><span class="n">csv</span> <span class="n">IMGT000035_genes</span><span class="o">.</span><span class="n">csv</span> <span class="n">forward</span> <span class="n">comparison_results</span>
</pre></div>
</div>
<p>The final argument specifies that the results should be put into files named <code class="docutils literal notranslate"><span class="pre">comparison_results</span></code>. Three files are produced with differing extensions: a summary graphic (.jpg), a text file listing differences
in sequences annotated as functional (.txt), and a detailed line-by-line comparison (.csv).</p>
<p><a class="reference external" href="https://github.com/williamdlees/digger/blob/main/tests/human/IGH/IMGT000035/comparison_results.jpg">comparison_results.jpg</a> summarises functional annotations found by digger and IMGT, according to which,
digger annotated as functional all genes so annotated by IMGT, and annotated an additional 2 V-genes and one J-gene as functional.</p>
<a class="reference internal image-reference" href="../_images/igh_results.jpg"><img alt="../_images/igh_results.jpg" src="../_images/igh_results.jpg" style="width: 600px;" /></a>
<p><a class="reference external" href="https://github.com/williamdlees/digger/blob/main/tests/human/IGH/IMGT000035/comparison_results.txt">comparison_results.txt</a> lists the differences in functional analysis in detail.
<a class="reference external" href="https://github.com/williamdlees/digger/blob/main/tests/human/IGH/IMGT000035/comparison_results_notes.txt">comparison_results_notes.txt</a> adds some commentary: of the two additional
V-genes annotated by digger as functional, one has unusual variations in the RSS, causing IMGT to annotate it as ORF. The other is annotated as ORF on the grounds that it has not been
seen rearranged. The additional J-gene is currently annotated by IMGT in the assembly as ORF, although it is listed in the IMGT gene table as functional.</p>
<p>There is not, at present, a clear set of universally accepted criteria for categorisation of functionality, and minor differences of this nature are to be expected. Over the next few years, we expect to see comparisons of genomic
sequencing of the loci with the expressed repertoire across multiple subjects, and this should allow a deeper understanding to develop. Overall, the comparison of digger results with the manually supervised curation
at IMGT shows a good level of agreement. It is possible that results may change from those noted here, as they are based on downloaded data which may be revised over time.</p>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this heading"></a></h2>
<p>Lefranc et al., 2015, IMGT®, the international ImMunoGeneTics information system® 25 years on. <em>Nucleic Acids Res.</em> <a class="reference external" href="https://doi.org/10.1093%2Fnar%2Fgku1056">doi: 10.1093/nar/gku1056</a>.</p>
<p>Watson et al., 2013, Complete haplotype sequence of the human immunoglobulin heavy-chain variable, diversity, and joining genes and characterization of allelic and copy-number variation. <em>Am J Hum Genet</em> <a class="reference external" href="http://dx.doi.org/10.1016/j.ajhg.2013.03.004">doi: 10.1016/j.ajhg.2013.03.004</a></p>
<p>Schneider et al., 2017, Evaluation of GRCh38 and de novo haploid genome assemblies demonstrates the enduring quality of the reference assembly. <em>Genome Research</em> <a class="reference external" href="https://genome.cshlp.org/content/27/5/849">doi: 10.1101/gr.213611.116</a></p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../news.html" class="btn btn-neutral float-left" title="Release Notes" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="rhesus_igh.html" class="btn btn-neutral float-right" title="Annotating the rhesus macaque IGH locus" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, William Lees.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>