<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Annotating the human IGHV locus &mdash; Digger 0.1 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Commandline Usage" href="../usage.html" />
    <link rel="prev" title="Release Notes" href="../news.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Digger
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../news.html">Release Notes</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Annotating the human IGHV locus</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#data">Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#annotating-the-assembly">Annotating the Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Commandline Usage</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Digger</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Annotating the human IGHV locus</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/examples/human_ighv.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <p>..human_ighv:</p>
<section id="annotating-the-human-ighv-locus">
<h1>Annotating the human IGHV locus<a class="headerlink" href="#annotating-the-human-ighv-locus" title="Permalink to this heading"></a></h1>
<p>The IGHV locus in the human reference assembly GRCh38.p12 has been annotated by IMGT. In this example, we will annotate with Digger and compare results.</p>
<section id="data">
<h2>Data<a class="headerlink" href="#data" title="Permalink to this heading"></a></h2>
<p>The locus and IMGT annotations can be downloaded and parsed with <a class="reference internal" href="../tools/parse_imgt_annotations.html#parse-imgt-annotations"><span class="std std-ref">parse_imgt_annotations</span></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">parse_imgt_annotations</span> \
    <span class="o">--</span><span class="n">save_sequence</span> <span class="n">IMGT000035</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="s2">&quot;http://www.imgt.org/ligmdb/view?format=IMGT&amp;id=IMGT000035&quot;</span> \
        <span class="n">IMGT000035_genes</span><span class="o">.</span><span class="n">csv</span> \
        <span class="n">IGH</span>
</pre></div>
</div>
<p>This will create two files: <code class="docutils literal notranslate"><span class="pre">IMGT000035.fasta</span></code>, containing the assembly sequence, and <code class="docutils literal notranslate"><span class="pre">IMGT000035_genes.csv</span></code>, containing the co-ordinates of the IMGT-annotated genes, and sequences of their flanking regions.</p>
<p>The human IGH reference set can be downloaded from IMGT with the <a class="reference external" href="https://williamdlees.github.io/receptor_utils/_build/html/introduction.html">receptor_utils</a> command <code class="docutils literal notranslate"><span class="pre">extract_refs</span></code> (receptor_utils is installed as part of Digger’s installation):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">extract_refs</span> <span class="o">-</span><span class="n">L</span> <span class="n">IGH</span> <span class="s2">&quot;Homo sapiens&quot;</span>
</pre></div>
</div>
<p>This creates the reference sets <code class="docutils literal notranslate"><span class="pre">Homo_sapiens_IGHV_gapped.fasta</span></code>, <code class="docutils literal notranslate"><span class="pre">Homo_sapiens_IGHV.fasta</span></code>, <code class="docutils literal notranslate"><span class="pre">Homo_sapiens_IGHD.fasta</span></code>, <code class="docutils literal notranslate"><span class="pre">Homo_sapiens_IGHJ.fasta</span></code>, <code class="docutils literal notranslate"><span class="pre">Homo_sapiens_CH.fasta</span></code></p>
<p>Lastly, we need to make a combined file containing all the IGH V, D and J reference genes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">cat</span> <span class="n">Homo_sapiens_IGHV</span><span class="o">.</span><span class="n">fasta</span> <span class="n">Homo_sapiens_IGHD</span><span class="o">.</span><span class="n">fasta</span> <span class="n">Homo_sapiens_IGHJ</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">&gt;</span> <span class="n">Homo_sapiens_IGHVDJ</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
</section>
<section id="annotating-the-assembly">
<h2>Annotating the Assembly<a class="headerlink" href="#annotating-the-assembly" title="Permalink to this heading"></a></h2>
<p>With the data in place, we can instruct digger to perform the annotation:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">digger</span> <span class="n">IMGT000035</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">-</span><span class="n">v_ref</span> <span class="n">Homo_sapiens_IGHV</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">-</span><span class="n">d_ref</span> <span class="n">Homo_sapiens_IGHD</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">-</span><span class="n">j_ref</span> <span class="n">Homo_sapiens_IGHJ</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">-</span><span class="n">v_ref_gapped</span> <span class="n">Homo_sapiens_IGHV_gapped</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">-</span><span class="n">ref</span> <span class="n">imgt</span><span class="p">,</span><span class="n">Homo_sapiens_IGHVDJ</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">-</span><span class="n">species</span> <span class="n">human</span> \
        <span class="o">-</span><span class="n">locus</span> <span class="n">IGH</span> \
        <span class="n">IMGT000035</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<p>the -v_ref, -d_ref, -j_ref and -V_ref_gapped arguments provide the reference assembly sequences.</p>
<p>The -ref argument instructs Digger to compare any sequences it identifies in the assembly with those in the combined reference
set. It will create columns in the output assembly containing the nearest reference sequence found, % identity and so on. These columns will be prefixed ‘imgt’ as specified in this argument. Multiple -ref arguments can be
passed to Digger, allowing comparison with multiple reference sets.</p>
<p>The -species argument tells Digger to use its internal position-weighted matrices for human loci, and -locus specifies the locus.</p>
<p>The output file will be <code class="docutils literal notranslate"><span class="pre">IMGT000035.csv</span></code>.</p>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this heading"></a></h2>
<p>Lefranc et al., 2015, IMGT®, the international ImMunoGeneTics information system® 25 years on. <em>Nucleic Acids Res.</em> <a class="reference external" href="https://doi.org/10.1093%2Fnar%2Fgku1056">doi: 10.1093/nar/gku1056</a>.</p>
<p>Watson et al., 2013, Complete haplotype sequence of the human immunoglobulin heavy-chain variable, diversity, and joining genes and characterization of allelic and copy-number variation. <em>Am J Hum Genet</em> <a class="reference external" href="http://dx.doi.org/10.1016/j.ajhg.2013.03.004">doi: 10.1016/j.ajhg.2013.03.004</a></p>
<p>Schneider et al., 2017, Evaluation of GRCh38 and de novo haploid genome assemblies demonstrates the enduring quality of the reference assembly. <em>Genome Research</em> <a class="reference external" href="https://genome.cshlp.org/content/27/5/849">doi: 10.1101/gr.213611.116</a></p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../news.html" class="btn btn-neutral float-left" title="Release Notes" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../usage.html" class="btn btn-neutral float-right" title="Commandline Usage" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, William Lees.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>