<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Annotating the rhesus macaque IGH locus &mdash; Digger 0.5.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Targeted Annotation" href="targeted_annotation.html" />
    <link rel="prev" title="Annotating the human IGH locus" href="human_igh.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Digger
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../overview.html#digger">digger</a></li>
<li class="toctree-l1"><a class="reference internal" href="../overview.html#dig-sequence">dig-sequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="../docker.html">Docker Image</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../news.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../news.html#changes-in-0-7-5">Changes in 0.7.5</a></li>
<li class="toctree-l1"><a class="reference internal" href="../news.html#changes-in-0-7-4">Changes in 0.7.4</a></li>
<li class="toctree-l1"><a class="reference internal" href="../news.html#changes-in-0-7-3">Changes in 0.7.3</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="human_igh.html">Annotating the human IGH locus</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Annotating the rhesus macaque IGH locus</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#data">Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#preparing-position-weighted-matrices">Preparing position-weighted matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="#annotating-the-assembly">Annotating the Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="#comparing-the-output-to-the-study-s-annotation">Comparing the output to the study’s annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="targeted_annotation.html">Targeted Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="additional_examples.html">Additional Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Commandline Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools/annotation.html">Anotation format</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Digger</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Annotating the rhesus macaque IGH locus</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/examples/rhesus_igh.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="annotating-the-rhesus-macaque-igh-locus">
<span id="rhesus-igh"></span><h1>Annotating the rhesus macaque IGH locus<a class="headerlink" href="#annotating-the-rhesus-macaque-igh-locus" title="Permalink to this heading"></a></h1>
<p>In this example we will see how to:</p>
<ul class="simple">
<li><p>Create position weight matrices from existing IMGT annotations</p></li>
<li><p>Use the underlying commands that digger calls, and understand how they might be useful when annotating multiple contigs or scaffolds</p></li>
</ul>
<p>IMGT has identified scaffolds in the 2006 rhesus macaque reference assembly, Mmul_051212, which lie within the IGH locus. Here we will bring them together in a single file and annotate them with motifs derived from the current reference assembly, rhemac10 (Mmul_10).
While this example is somewhat artificial, in that the scaffolds could equally well be annotated using the single digger command, the example  serves to show how the individual commands in the package can be used. This provides some additional flexibility, for example in tuning the blast searches, and also illustrates how they work together.
The comparison with IMGT’s annotation of Mmul_051212, and a script to reproduce this example using the steps below, can be found <a class="reference external" href="https://github.com/williamdlees/digger/tree/main/tests/rhesus_macaque/IGH/Mmul_51212">in digger’s Git repository</a>.</p>
<section id="data">
<h2>Data<a class="headerlink" href="#data" title="Permalink to this heading"></a></h2>
<p>As in the previous example, the rhesus IGH germline reference set can be downloaded from IMGT with the <a class="reference external" href="https://williamdlees.github.io/receptor_utils/_build/html/introduction.html">receptor_utils</a> command
<code class="docutils literal notranslate"><span class="pre">extract_refs</span></code> (receptor_utils is installed as part of digger’s installation). However, the rhesus IG gapped V-genes provided by IMGT contain additional inserted codons relative to
the conventional IMGT alignment. As digger (alongside other tools) expects the conventional alignment, a further step is needed to realign the gapped sequences, using the receptor_utils
tool <code class="docutils literal notranslate"><span class="pre">fix_macaque_gaps</span></code>. The following commands will prepare the reference data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">extract_refs</span> <span class="o">-</span><span class="n">L</span> <span class="n">IGH</span> <span class="s2">&quot;Macaca mulatta&quot;</span>
<span class="o">&gt;</span> <span class="n">fix_macaque_gaps</span> <span class="n">Macaca_mulatta_IGHV_gapped</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="n">Macaca_mulatta_IGHV_gapped_fixed</span><span class="o">.</span><span class="n">fasta</span> <span class="n">IGH</span>
<span class="o">&gt;</span> <span class="n">cat</span> <span class="n">Macaca_mulatta_IGHV</span><span class="o">.</span><span class="n">fasta</span> <span class="n">Macaca_mulatta_IGHD</span><span class="o">.</span><span class="n">fasta</span> <span class="n">Macaca_mulatta_IGHJ</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">&gt;</span> <span class="n">Macaca_mulatta_IGHVDJ</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>The Mmul_51212 scaffolds can be downloaded from IMGT as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">parse_imgt_annotations</span> <span class="o">--</span><span class="n">save_sequence</span> <span class="n">NW_001157919</span><span class="o">.</span><span class="n">fasta</span> \
   <span class="s2">&quot;https://www.imgt.org/ligmdb/view.action?format=IMGT&amp;id=NW_001157919&quot;</span> \
       <span class="n">NW_001157919_genes</span><span class="o">.</span><span class="n">csv</span> <span class="n">IGH</span>
<span class="o">&gt;</span> <span class="n">parse_imgt_annotations</span> <span class="o">--</span><span class="n">save_sequence</span> <span class="n">NW_001122023</span><span class="o">.</span><span class="n">fasta</span> \
   <span class="s2">&quot;https://www.imgt.org/ligmdb/view.action?format=IMGT&amp;id=NW_001122023&quot;</span> \
   <span class="n">NW_001122023_genes</span><span class="o">.</span><span class="n">csv</span> <span class="n">IGH</span>
<span class="o">&gt;</span> <span class="n">parse_imgt_annotations</span> <span class="o">--</span><span class="n">save_sequence</span> <span class="n">NW_001122024</span><span class="o">.</span><span class="n">fasta</span> \
   <span class="s2">&quot;https://www.imgt.org/ligmdb/view.action?format=IMGT&amp;id=NW_001122024&quot;</span> \
       <span class="n">NW_001122024_genes</span><span class="o">.</span><span class="n">csv</span> <span class="n">IGH</span>
<span class="o">&gt;</span> <span class="n">parse_imgt_annotations</span> <span class="o">--</span><span class="n">save_sequence</span> <span class="n">NW_001121239</span><span class="o">.</span><span class="n">fasta</span> \
   <span class="s2">&quot;https://www.imgt.org/ligmdb/view.action?format=IMGT&amp;id=NW_001121239&quot;</span> \
       <span class="n">NW_001121239_genes</span><span class="o">.</span><span class="n">csv</span> <span class="n">IGH</span>
<span class="o">&gt;</span> <span class="n">parse_imgt_annotations</span> <span class="o">--</span><span class="n">save_sequence</span> <span class="n">NW_001121240</span><span class="o">.</span><span class="n">fasta</span> \
   <span class="s2">&quot;https://www.imgt.org/ligmdb/view.action?format=IMGT&amp;id=NW_001121240&quot;</span> \
       <span class="n">NW_001121240_genes</span><span class="o">.</span><span class="n">csv</span> <span class="n">IGH</span>

<span class="o">&gt;</span><span class="n">cat</span> <span class="n">NW_001157919</span><span class="o">.</span><span class="n">fasta</span> <span class="n">NW_001122023</span><span class="o">.</span><span class="n">fasta</span> <span class="n">NW_001122024</span><span class="o">.</span><span class="n">fasta</span> \
       <span class="n">NW_001121239</span><span class="o">.</span><span class="n">fasta</span> <span class="n">NW_001121240</span><span class="o">.</span><span class="n">fasta</span> <span class="o">&gt;</span> <span class="n">Mmul_051212</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
</section>
<section id="preparing-position-weighted-matrices">
<h2>Preparing position-weighted matrices<a class="headerlink" href="#preparing-position-weighted-matrices" title="Permalink to this heading"></a></h2>
<p>Digger already has PWMs for rhesus IGH, but for the purpose of this example, we will create a set using the features listed in IMGT’s annotation of the rhemac10 IGH locus, which
has the IMGT accession number IMGT000064. The following commands download the annotation, determine the features, and calculate the PWMs from
features of functional annotations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">mkdir</span> <span class="n">motifs</span>
<span class="o">&gt;</span> <span class="n">cd</span> <span class="n">motifs</span>
<span class="o">&gt;</span> <span class="n">parse_imgt_annotations</span> \
        <span class="s2">&quot;http://www.imgt.org/ligmdb/view?format=IMGT&amp;id=IMGT000064&quot;</span> \
        <span class="n">IMGT000064_genes</span><span class="o">.</span><span class="n">csv</span> <span class="n">IGH</span>
<span class="o">&gt;</span> <span class="n">calc_motifs</span> <span class="n">IGH</span> <span class="n">IMGT000064_genes</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">calc_motifs</span></code> will create 10 motif files in the directory. Additionally, it prepares <code class="docutils literal notranslate"><span class="pre">motif_params.json</span></code>, which contains some further locus-specific motif parameters.
These are descrbed further in <a class="reference internal" href="../tools/calc_motifs.html#calc-motifs"><span class="std std-ref">calc_motifs</span></a>.</p>
<p>The motifs directory may optionally contain a FASTA file <code class="docutils literal notranslate"><span class="pre">conserved_motifs.fasta</span></code> defining strongly-conserved nucleotides in the RSS and leader. Only those features
with conserved nucleotides need to be listed in the file. The names follow the filenames used for the PWMs.
The following sequences were derived from Figure 3 of Ngoune et al. (2022) and will be used in this example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">V</span><span class="o">-</span><span class="n">HEPTAMER</span>
<span class="n">CAC</span><span class="o">---</span><span class="n">G</span>
<span class="o">&gt;</span><span class="n">V</span><span class="o">-</span><span class="n">NONAMER</span>
<span class="o">-----</span><span class="n">AACC</span>
<span class="o">&gt;</span><span class="mi">5</span><span class="s1">&#39;D-HEPTAMER</span>
<span class="o">----</span><span class="n">GTG</span>
<span class="o">&gt;</span><span class="mi">5</span><span class="s1">&#39;D-NONAMER</span>
<span class="o">---</span><span class="n">T</span><span class="o">-----</span>
<span class="o">&gt;</span><span class="mi">3</span><span class="s1">&#39;D-HEPTAMER</span>
<span class="n">C</span><span class="o">-</span><span class="n">C</span><span class="o">---</span><span class="n">G</span>
<span class="o">&gt;</span><span class="mi">3</span><span class="s1">&#39;D-NONAMER</span>
<span class="o">-</span><span class="n">C</span><span class="o">----</span><span class="n">A</span><span class="o">--</span>
<span class="o">&gt;</span><span class="n">J</span><span class="o">-</span><span class="n">HEPTAMER</span>
<span class="n">C</span><span class="o">--</span><span class="n">TGTG</span>
<span class="o">&gt;</span><span class="n">J</span><span class="o">-</span><span class="n">NONAMER</span>
<span class="o">-</span><span class="n">GTT</span><span class="o">--</span><span class="n">TG</span><span class="o">-</span>
</pre></div>
</div>
<p>Again, this file is provided for download at the location provided near the top of this example.</p>
<p>While the presence or absence of conserved nucleotides can be a useful guide to the likely functionality of a sequence, please bear in mind that it is a guide only:
exceptions can be expected, particularly where the definitions have been built on limited data.</p>
</section>
<section id="annotating-the-assembly">
<h2>Annotating the Assembly<a class="headerlink" href="#annotating-the-assembly" title="Permalink to this heading"></a></h2>
<p>The digger command is not able to handle a FASTA file containing multiple contigs, so we will call the underlying tools directly. We start by creating the blast databases and querying against the assembly,
using the reference genes determined in the study:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">makeblastdb</span> <span class="o">-</span><span class="ow">in</span> <span class="n">Macaca_mulatta_IGHV</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">dbtype</span> <span class="n">nucl</span>
<span class="o">&gt;</span> <span class="n">makeblastdb</span> <span class="o">-</span><span class="ow">in</span> <span class="n">Macaca_mulatta_IGHD</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">dbtype</span> <span class="n">nucl</span>
<span class="o">&gt;</span> <span class="n">makeblastdb</span> <span class="o">-</span><span class="ow">in</span> <span class="n">Macaca_mulatta_IGHJ</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">dbtype</span> <span class="n">nucl</span>

<span class="o">&gt;</span> <span class="n">blastn</span> <span class="o">-</span><span class="n">db</span> <span class="n">Macaca_mulatta_IGHV</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">query</span> <span class="n">Mmul_051212</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">out</span> <span class="n">mmul_IGHV</span><span class="o">.</span><span class="n">out</span> \
   <span class="o">-</span><span class="n">outfmt</span> <span class="mi">7</span> <span class="o">-</span><span class="n">gapopen</span> <span class="mi">5</span> <span class="o">-</span><span class="n">gapextend</span> <span class="mi">5</span> <span class="o">-</span><span class="n">penalty</span> <span class="o">-</span><span class="mi">1</span> <span class="o">-</span><span class="n">word_size</span> <span class="mi">11</span>
<span class="o">&gt;</span> <span class="n">blastn</span> <span class="o">-</span><span class="n">db</span> <span class="n">Macaca_mulatta_IGHD</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">query</span> <span class="n">Mmul_051212</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">out</span> <span class="n">mmul_IGHD</span><span class="o">.</span><span class="n">out</span> \
   <span class="o">-</span><span class="n">outfmt</span> <span class="mi">7</span> <span class="o">-</span><span class="n">gapopen</span> <span class="mi">5</span> <span class="o">-</span><span class="n">gapextend</span> <span class="mi">5</span> <span class="o">-</span><span class="n">penalty</span> <span class="o">-</span><span class="mi">1</span> <span class="o">-</span><span class="n">word_size</span> <span class="mi">7</span> <span class="o">-</span><span class="n">evalue</span> <span class="mi">100</span>
<span class="o">&gt;</span> <span class="n">blastn</span> <span class="o">-</span><span class="n">db</span> <span class="n">Macaca_mulatta_IGHJ</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">query</span> <span class="n">Mmul_051212</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">out</span> <span class="n">mmul_IGHJ</span><span class="o">.</span><span class="n">out</span> \
   <span class="o">-</span><span class="n">outfmt</span> <span class="mi">7</span> <span class="o">-</span><span class="n">gapopen</span> <span class="mi">5</span> <span class="o">-</span><span class="n">gapextend</span> <span class="mi">5</span> <span class="o">-</span><span class="n">penalty</span> <span class="o">-</span><span class="mi">1</span> <span class="o">-</span><span class="n">word_size</span> <span class="mi">7</span>
</pre></div>
</div>
<p>Note that a higher evalue is used for the D genes, as they can be quite short.</p>
<p>Next we call <code class="docutils literal notranslate"><span class="pre">blastresults_to_csv</span></code> to convert to a more convenient format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">blastresults_to_csv</span> <span class="n">mmul_IGHV</span><span class="o">.</span><span class="n">out</span> <span class="n">mmul_ighvdj_</span>
<span class="o">&gt;</span> <span class="n">blastresults_to_csv</span> <span class="n">mmul_IGHD</span><span class="o">.</span><span class="n">out</span> <span class="n">mmul_ighvdj_</span> <span class="o">-</span><span class="n">a</span>
<span class="o">&gt;</span> <span class="n">blastresults_to_csv</span> <span class="n">mmul_IGHJ</span><span class="o">.</span><span class="n">out</span> <span class="n">mmul_ighvdj_</span> <span class="o">-</span><span class="n">a</span>
</pre></div>
</div>
<p>The commands instruct the tool to create merged files containing V,D and J hits. This is achieved by specifying the same prefix on each command <code class="docutils literal notranslate"><span class="pre">(mmul_ighvdj_)</span></code> and using the <code class="docutils literal notranslate"><span class="pre">-a</span></code> (append) option.
The records created by blastn contain the name of the contig in which a hit was found. <code class="docutils literal notranslate"><span class="pre">blastresults_to_csv</span></code> will create one file per contig. The names contain the ID of the contig in
<code class="docutils literal notranslate"><span class="pre">Mmul_051212.fasta</span></code>, except that they are modified where necessary to ensure file system compatibility.</p>
<p>We now call <code class="docutils literal notranslate"><span class="pre">find_alignments</span></code> to process the annotations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">find_alignments</span> <span class="n">Macaca_mulatta_IGHVDJ</span><span class="o">.</span><span class="n">fasta</span> \
       <span class="n">Mmul_051212</span><span class="o">.</span><span class="n">fasta</span> \
       <span class="s2">&quot;mmul_ighvdj_nw_*.csv&quot;</span> \
       <span class="o">-</span><span class="n">ref</span> <span class="n">imgt</span><span class="p">,</span><span class="n">Macaca_mulatta_IGHVDJ</span><span class="o">.</span><span class="n">fasta</span> \
       <span class="o">-</span><span class="n">align</span> <span class="n">Macaca_mulatta_IGHV_gapped_fixed</span><span class="o">.</span><span class="n">fasta</span> \
       <span class="o">-</span><span class="n">motif_dir</span> <span class="n">motifs</span> \
       <span class="n">Mmul_051212</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<p>Note that the third argument, <code class="docutils literal notranslate"><span class="pre">&quot;mmul_ighvdj_nw_*.csv&quot;</span></code>, contains a wildcard that will match all the files produced in the previous step. It is quoted to avoid expansion by the shell.
V-genes in the annotation will be annotated and gapped using the IMGT set as a template (with fixed gaps).
<code class="docutils literal notranslate"><span class="pre">find_alignments</span></code> will attempt to deduce the sense in which to annotate each segment. This is helpful in this case as the contigs vary in their orientation.  Note that we are
specifying the location of the motifs directory created in the previous step rather than the species and locus, which would cause digger to use the built-in tables.</p>
</section>
<section id="comparing-the-output-to-the-study-s-annotation">
<h2>Comparing the output to the study’s annotation<a class="headerlink" href="#comparing-the-output-to-the-study-s-annotation" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">compare_annotations</span></code> is not capable of handling the output from multiple sequences in the same file, so unfortunately we need to split the results up for the comparison:</p>
<blockquote>
<div><p>&gt; head -n 1 Mmul_051212.csv &gt; mmul_header.csv</p>
<p>&gt; cp mmul_header.csv NW_001157919_digger.csv
&gt; grep NW_001157919 Mmul_051212.csv &gt;&gt; NW_001157919_digger.csv</p>
<p>&gt; cp mmul_header.csv NW_001122023_digger.csv
&gt; grep NW_001122023 Mmul_051212.csv &gt;&gt; NW_001122023_digger.csv</p>
<p>&gt; cp mmul_header.csv NW_001122024_digger.csv
&gt; grep NW_001122024 Mmul_051212.csv &gt;&gt; NW_001122024_digger.csv</p>
<p>&gt; cp mmul_header.csv NW_001121239_digger.csv
&gt; grep NW_001121239 Mmul_051212.csv &gt;&gt; NW_001121239_digger.csv</p>
<p>&gt; cp mmul_header.csv NW_001121240_digger.csv
&gt; grep NW_001121240 Mmul_051212.csv &gt;&gt; NW_001121240_digger.csv</p>
<p>&gt; compare_annotations NW_001157919_digger.csv NW_001157919_genes.csv forward NW_001157919_comp
&gt; compare_annotations NW_001122023_digger.csv NW_001122023_genes.csv forward NW_001122023_comp
&gt; compare_annotations NW_001122024_digger.csv NW_001122024_genes.csv forward NW_001122024_comp
&gt; compare_annotations NW_001121239_digger.csv NW_001121239_genes.csv forward NW_001121239_comp
&gt; compare_annotations NW_001121240_digger.csv NW_001121240_genes.csv forward NW_001121240_comp</p>
</div></blockquote>
<p>Scaffold-by-scaffold comparisons are provided in <a class="reference external" href="https://github.com/williamdlees/digger/tree/main/tests/rhesus_macaque/IGH/Mmul_051212">Github</a>.
and an overall comparison is provided <a class="reference external" href="https://github.com/williamdlees/digger/tree/main/tests/rhesus_macaque/IGH/Mmul_051212/comparison_notes.txt">here</a>.</p>
<p>Digger identified a total of 13 potentially functional D-genes not annotated by IMGT, across four of the five scaffolds, while IMGT annotated D-genes only in NW_001121239. The macaque IGHD genes are known
to occupy a small, distinct, region towards the 3’ end of the IGH locus. It would therefore be reasonable to expect them to be located in a single scaffold, and to be
distinct from the V-genes. However, given the sequencing technology available for sequencing and assembly when the scaffolds were created, and bearing
in mind the short length of the D-genes, it is possible that the D-locus was not correctly assembled. Another reason for suspecting this is that two of the D-sequences
identified by Digger are extremely short, at 3nt and 1nt, and yet appear to be flanked by functional RSS. In contrast, in an
<a class="reference external" href="https://github.com/williamdlees/digger/tree/main/tests/rhesus_macaque/IGH/IMGT000064">annotation of the rhemac10 assembly</a>, Digger identified only one D-gene
not annotated by IMGT (this was also outside the D locus).</p>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this heading"></a></h2>
<p>Ngoune et al., 2022, IMGT® Biocuration and Analysis of the Rhesus Monkey IG Loci. <em>Vaccines</em> <a class="reference external" href="https://www.mdpi.com/2076-393X/10/3/394#">doi: 10.3390/vaccines10030394</a>.</p>
<p>Warren et al., 2020, Sequence Diversity Analyses of an Improved Rhesus Macaque Genome Enhance Its Biomedical Utility. <em>Science</em> <a class="reference external" href="https://doi.org/10.1126/science.abc6617">doi: 10.1126/science.abc6617</a>.</p>
<p>Gibbs et al., 2007, Evolutionary and biomedical insights from the rhesus macaque genome. <em>Science</em> <a class="reference external" href="https://doi.org/10.1126/science.1139247">doi: 10.1126/science.1139247</a>.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="human_igh.html" class="btn btn-neutral float-left" title="Annotating the human IGH locus" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="targeted_annotation.html" class="btn btn-neutral float-right" title="Targeted Annotation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, William Lees.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>