%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax
\ifdefined\pdfimageresolution
    \pdfimageresolution= \numexpr \dimexpr1in\relax/\sphinxpxdimen\relax
\fi
%% let collapsible pdf bookmarks panel have high depth per default
\PassOptionsToPackage{bookmarksdepth=5}{hyperref}


\PassOptionsToPackage{warn}{textcomp}
\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
% support both utf8 and utf8x syntaxes
  \ifdefined\DeclareUnicodeCharacterAsOptional
    \def\sphinxDUC#1{\DeclareUnicodeCharacter{"#1}}
  \else
    \let\sphinxDUC\DeclareUnicodeCharacter
  \fi
  \sphinxDUC{00A0}{\nobreakspace}
  \sphinxDUC{2500}{\sphinxunichar{2500}}
  \sphinxDUC{2502}{\sphinxunichar{2502}}
  \sphinxDUC{2514}{\sphinxunichar{2514}}
  \sphinxDUC{251C}{\sphinxunichar{251C}}
  \sphinxDUC{2572}{\textbackslash}
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}



\usepackage{tgtermes}
\usepackage{tgheros}
\renewcommand{\ttdefault}{txtt}



\usepackage[Bjarne]{fncychap}
\usepackage{sphinx}

\fvset{fontsize=auto}
\usepackage{geometry}


% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}

\addto\captionsenglish{\renewcommand{\contentsname}{Getting Started}}

\usepackage{sphinxmessages}
\setcounter{tocdepth}{0}



\title{Digger}
\date{Apr 06, 2023}
\release{0.1}
\author{William Lees}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex
\begin{document}

\ifdefined\shorthandoff
  \ifnum\catcode`\=\string=\active\shorthandoff{=}\fi
  \ifnum\catcode`\"=\active\shorthandoff{"}\fi
\fi

\pagestyle{empty}
\sphinxmaketitle
\pagestyle{plain}
\sphinxtableofcontents
\pagestyle{normal}
\phantomsection\label{\detokenize{index::doc}}


\sphinxAtStartPar
Digger is a toolkit for the automatic annotation of unrearranged V,D and J genes in B\sphinxhyphen{} and T\sphinxhyphen{} cell immunoglobulin receptor genomic loci. It can be used to annotate both entire assemblies or large fragments of a locus,
or small fragments. It is designed for use with entire gene sequences, including leader and RSS (V\sphinxhyphen{}gene UTRs are not required, and are currently not annotated).

\sphinxstepscope


\chapter{Overview}
\label{\detokenize{overview:overview}}\label{\detokenize{overview:overview-label}}\label{\detokenize{overview::doc}}

\section{Scope and Features}
\label{\detokenize{overview:scope-and-features}}
\sphinxAtStartPar
Digger identifies immunoglobulin receptor germline sequences in a genomic sequence or assembly. It requires two inputs: the sequence or assembly, and a set of reference genes to use as a starting point. An existing
reference set for the species under study is an ideal starting point, but if one is not available, good results can be obtained by using the reference set for another species, preferably one that is
reasonably closely related. The package is intended to be straightforward to use, but some familiarity with command\sphinxhyphen{}line tools is expected.

\sphinxAtStartPar
Digger starts by searching the sequence for approximate matches to sequences in the reference set. Where these are found, the match is extended to the full length of the matched sequence. A window
at either end is then checked for the expected flanking sequences (e.g. leader, RSS). These are identified by means of position weight matrices (PWMs). The flanking sequences have a well\sphinxhyphen{}established
‘canonical’ form that is conserved between species. Digger contains PWMs for human and rhesus macaque IG loci, and these can be used as a starting point for other species. However, as some variation
is observed between species, Digger also contains tools for deriving tailored PWMs for a species of interest. These can be obtained from an existing annotation of the locus, or from an initial
annotation conducted with the human or rhesus PWMs. Digger’s built\sphinxhyphen{}in PWM’s are based on analysis of selected IMGT annotations.


\section{Pipeline}
\label{\detokenize{overview:pipeline}}
\noindent\sphinxincludegraphics[width=600\sphinxpxdimen]{{pipeline}.jpg}

\sphinxAtStartPar
A brief summary of the tools contained in the package is given below. In most cases, annotation can be performed with the {\hyperref[\detokenize{tools/digger:digger}]{\sphinxcrossref{\DUrole{std,std-ref}{digger}}}} tool. This will call subsidiary tools as necessary. Please refer
to the Examples section for more information.


\begin{savenotes}\sphinxattablestart
\sphinxthistablewithglobalstyle
\centering
\phantomsection\label{\detokenize{overview:featuretable}}\nobreak
\begin{tabular}[t]{|\X{25}{100}|\X{75}{100}|}
\sphinxtoprule
\sphinxstyletheadfamily 
\sphinxAtStartPar
Tool
&\sphinxstyletheadfamily 
\sphinxAtStartPar
Description
\\
\sphinxmidrule
\sphinxtableatstartofbodyhook
\sphinxAtStartPar
{\hyperref[\detokenize{tools/blastresults_to_csv:blastresults-to-csv}]{\sphinxcrossref{\DUrole{std,std-ref}{blastresults\_to\_csv}}}}
&
\sphinxAtStartPar
Convert the result of a BLAST search to a simple tabular format
\\
\sphinxhline
\sphinxAtStartPar
{\hyperref[\detokenize{tools/calc_motifs:calc-motifs}]{\sphinxcrossref{\DUrole{std,std-ref}{calc\_motifs}}}}
&
\sphinxAtStartPar
Determine PWMs from a list of gene features
\\
\sphinxhline
\sphinxAtStartPar
{\hyperref[\detokenize{tools/compare_annotations:compare-annotations}]{\sphinxcrossref{\DUrole{std,std-ref}{compare\_annotations}}}}
&
\sphinxAtStartPar
Compare two sets of annotations
\\
\sphinxhline
\sphinxAtStartPar
{\hyperref[\detokenize{tools/digger:digger}]{\sphinxcrossref{\DUrole{std,std-ref}{digger}}}}
&
\sphinxAtStartPar
Run the annotation pipeline
\\
\sphinxhline
\sphinxAtStartPar
{\hyperref[\detokenize{tools/find_alignments:find-alignments}]{\sphinxcrossref{\DUrole{std,std-ref}{find\_alignments}}}}
&
\sphinxAtStartPar
Produce the annotation file, given tabulated BLAST results and other inputs
\\
\sphinxhline
\sphinxAtStartPar
{\hyperref[\detokenize{tools/parse_imgt_annotations:parse-imgt-annotations}]{\sphinxcrossref{\DUrole{std,std-ref}{parse\_imgt\_annotations}}}}
&
\sphinxAtStartPar
Download and parse an annotation file from IMGT
\\
\sphinxbottomrule
\end{tabular}
\sphinxtableafterendhook\par
\sphinxattableend\end{savenotes}

\sphinxstepscope


\chapter{Installation}
\label{\detokenize{install:installation}}\label{\detokenize{install:install}}\label{\detokenize{install::doc}}
\sphinxAtStartPar
The latest stable version of Digger may be downloaded from \sphinxhref{https://pypi.python.org/pypi/receptor-digger}{PyPI}. Digger requires Python 3.9 or above.

\sphinxAtStartPar
Development versions are available from \sphinxhref{https://github.com/williamdlees/digger}{GitHub}.

\sphinxAtStartPar
Digger requires \sphinxhref{https://www.ncbi.nlm.nih.gov/books/NBK279690/}{BLAST} to be installed.
No BLAST databases are needed: just the executable. Once BLAST has been installed, please verify by typing \sphinxtitleref{blastn \textendash{}help} at the command line: if everything is ok, it should provide
usage instructions.

\sphinxAtStartPar
If you encounter the error \sphinxcode{\sphinxupquote{Cannot allocate memory}}, this is coming from BLAST makeblastdb. Please set the environment variable \sphinxcode{\sphinxupquote{BLASTDB\_LMDB\_MAP\_SIZE=100000000}}.

\sphinxAtStartPar
The easiest way to install Digger itself is with pip:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}} \PYG{n}{pip} \PYG{n}{install} \PYG{n}{receptor}\PYG{o}{\PYGZhy{}}\PYG{n}{digger} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{user}
\end{sphinxVerbatim}

\sphinxAtStartPar
Digger requires the Python packages \sphinxhref{https://pypi.org/project/receptor-utils/}{receptor\sphinxhyphen{}utils} and \sphinxhref{https://pypi.org/project/biopython/}{biopython}. These should be installed automatically by pip.

\sphinxstepscope


\chapter{Release Notes}
\label{\detokenize{news:release-notes}}\label{\detokenize{news:news}}\label{\detokenize{news::doc}}

\section{Version 0.5: April 2023}
\label{\detokenize{news:version-0-5-april-2023}}
\sphinxAtStartPar
First public version.

\sphinxstepscope


\chapter{Annotating the human IGH locus}
\label{\detokenize{examples/human_igh:annotating-the-human-igh-locus}}\label{\detokenize{examples/human_igh:human-igh}}\label{\detokenize{examples/human_igh::doc}}
\sphinxAtStartPar
The IGHV locus in the human reference assembly GRCh38.p12 has been annotated by IMGT. In this example, we will annotate with Digger and compare results.
The comparison, and a script to reproduce it using the steps below, can be found \sphinxhref{https://github.com/williamdlees/digger/tree/main/tests/human/IGH/IMGT000035}{in digger’s Git repository}.


\section{Data}
\label{\detokenize{examples/human_igh:data}}
\sphinxAtStartPar
The locus and IMGT annotations can be downloaded and parsed with {\hyperref[\detokenize{tools/parse_imgt_annotations:parse-imgt-annotations}]{\sphinxcrossref{\DUrole{std,std-ref}{parse\_imgt\_annotations}}}}:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}} \PYG{n}{parse\PYGZus{}imgt\PYGZus{}annotations} \PYGZbs{}
    \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{save\PYGZus{}sequence} \PYG{n}{IMGT000035}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
        \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{http://www.imgt.org/ligmdb/view?format=IMGT\PYGZam{}id=IMGT000035}\PYG{l+s+s2}{\PYGZdq{}} \PYGZbs{}
        \PYG{n}{IMGT000035\PYGZus{}genes}\PYG{o}{.}\PYG{n}{csv} \PYGZbs{}
        \PYG{n}{IGH}
\end{sphinxVerbatim}

\sphinxAtStartPar
This will create two files: \sphinxcode{\sphinxupquote{IMGT000035.fasta}}, containing the assembly sequence, and \sphinxcode{\sphinxupquote{IMGT000035\_genes.csv}}, containing the co\sphinxhyphen{}ordinates of the IMGT\sphinxhyphen{}annotated genes, and sequences of their flanking regions.

\sphinxAtStartPar
The human IGH reference set can be downloaded from IMGT with the \sphinxhref{https://williamdlees.github.io/receptor\_utils/\_build/html/introduction.html}{receptor\_utils} command \sphinxcode{\sphinxupquote{extract\_refs}} (receptor\_utils is installed as part of Digger’s installation):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}} \PYG{n}{extract\PYGZus{}refs} \PYG{o}{\PYGZhy{}}\PYG{n}{L} \PYG{n}{IGH} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Homo sapiens}\PYG{l+s+s2}{\PYGZdq{}}
\end{sphinxVerbatim}

\sphinxAtStartPar
This creates the reference sets \sphinxcode{\sphinxupquote{Homo\_sapiens\_IGHV\_gapped.fasta}}, \sphinxcode{\sphinxupquote{Homo\_sapiens\_IGHV.fasta}}, \sphinxcode{\sphinxupquote{Homo\_sapiens\_IGHD.fasta}}, \sphinxcode{\sphinxupquote{Homo\_sapiens\_IGHJ.fasta}}, \sphinxcode{\sphinxupquote{Homo\_sapiens\_CH.fasta}}

\sphinxAtStartPar
Lastly, we need to make a combined file containing all the IGH V, D and J reference genes:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}} \PYG{n}{cat} \PYG{n}{Homo\PYGZus{}sapiens\PYGZus{}IGHV}\PYG{o}{.}\PYG{n}{fasta} \PYG{n}{Homo\PYGZus{}sapiens\PYGZus{}IGHD}\PYG{o}{.}\PYG{n}{fasta} \PYG{n}{Homo\PYGZus{}sapiens\PYGZus{}IGHJ}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
    \PYG{o}{\PYGZgt{}} \PYG{n}{Homo\PYGZus{}sapiens\PYGZus{}IGHVDJ}\PYG{o}{.}\PYG{n}{fasta}
\end{sphinxVerbatim}


\section{Annotating the Assembly}
\label{\detokenize{examples/human_igh:annotating-the-assembly}}
\sphinxAtStartPar
With the data in place, we can instruct {\hyperref[\detokenize{tools/digger:digger}]{\sphinxcrossref{\DUrole{std,std-ref}{digger}}}} to perform the annotation:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}} \PYG{n}{digger} \PYG{n}{IMGT000035}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
    \PYG{o}{\PYGZhy{}}\PYG{n}{v\PYGZus{}ref} \PYG{n}{Homo\PYGZus{}sapiens\PYGZus{}IGHV}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
        \PYG{o}{\PYGZhy{}}\PYG{n}{d\PYGZus{}ref} \PYG{n}{Homo\PYGZus{}sapiens\PYGZus{}IGHD}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
        \PYG{o}{\PYGZhy{}}\PYG{n}{j\PYGZus{}ref} \PYG{n}{Homo\PYGZus{}sapiens\PYGZus{}IGHJ}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
        \PYG{o}{\PYGZhy{}}\PYG{n}{v\PYGZus{}ref\PYGZus{}gapped} \PYG{n}{Homo\PYGZus{}sapiens\PYGZus{}IGHV\PYGZus{}gapped}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
        \PYG{o}{\PYGZhy{}}\PYG{n}{ref} \PYG{n}{imgt}\PYG{p}{,}\PYG{n}{Homo\PYGZus{}sapiens\PYGZus{}IGHVDJ}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
        \PYG{o}{\PYGZhy{}}\PYG{n}{species} \PYG{n}{human} \PYGZbs{}
        \PYG{o}{\PYGZhy{}}\PYG{n}{locus} \PYG{n}{IGH} \PYGZbs{}
        \PYG{n}{IMGT000035}\PYG{o}{.}\PYG{n}{csv}
\end{sphinxVerbatim}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{\sphinxhyphen{}v\_ref, \sphinxhyphen{}d\_ref, \sphinxhyphen{}j\_ref, \sphinxhyphen{}v\_ref\_gapped}} provide the reference assembly sequences.

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{\sphinxhyphen{}ref imgt,Homo\_sapiens\_IGHVDJ.fasta}} instructs Digger to compare any sequences it identifies in the assembly with those in the combined reference
set. It will create columns in the output assembly containing the nearest reference sequence found, \% identity and so on. These columns will be prefixed \sphinxcode{\sphinxupquote{imgt}} as specified in this argument. Multiple \sphinxcode{\sphinxupquote{\sphinxhyphen{}ref}} arguments can be
passed, allowing comparison with multiple reference sets.

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{\sphinxhyphen{}species}} tells Digger to use its internal position\sphinxhyphen{}weighted matrices for human loci, and \sphinxcode{\sphinxupquote{\sphinxhyphen{}locus}} specifies the locus.

\sphinxAtStartPar
The output file will be \sphinxcode{\sphinxupquote{IMGT000035.csv}}.

\sphinxAtStartPar
Digger will summarise progress as it runs. It will call the following tools:
\begin{itemize}
\item {} 
\sphinxAtStartPar
makeblastdb to create BLAST databases for the reference genes

\item {} 
\sphinxAtStartPar
blastn to run these databases against the reference sequence

\item {} 
\sphinxAtStartPar
{\hyperref[\detokenize{tools/blastresults_to_csv:blastresults-to-csv}]{\sphinxcrossref{\DUrole{std,std-ref}{blastresults\_to\_csv}}}} to convert the BLAST output to a simpler format

\item {} 
\sphinxAtStartPar
{\hyperref[\detokenize{tools/find_alignments:find-alignments}]{\sphinxcrossref{\DUrole{std,std-ref}{find\_alignments}}}} to annotate gene alignments found in the assembly

\end{itemize}


\section{Comparing the output to IMGT’s annotation}
\label{\detokenize{examples/human_igh:comparing-the-output-to-imgt-s-annotation}}
\sphinxAtStartPar
{\hyperref[\detokenize{tools/compare_annotations:compare-annotations}]{\sphinxcrossref{\DUrole{std,std-ref}{compare\_annotations}}}} compares the output of a digger run with a summary annotation file. Here we use it to compare the results of digger’s annotation with IMGT’s:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}} \PYG{n}{compare\PYGZus{}annotations} \PYG{n}{IMGT000035}\PYG{o}{.}\PYG{n}{csv} \PYG{n}{IMGT000035\PYGZus{}genes}\PYG{o}{.}\PYG{n}{csv} \PYG{n}{forward} \PYG{n}{comparison\PYGZus{}results}
\end{sphinxVerbatim}

\sphinxAtStartPar
The final argument specifies that the results should be put into files named \sphinxcode{\sphinxupquote{comparison\_results}}. Three files are produced with differing extensions: a summary graphic (.jpg), a text file listing differences
in sequences annotated as functional (.txt), and a detailed line\sphinxhyphen{}by\sphinxhyphen{}line comparison (.csv).

\sphinxAtStartPar
\sphinxhref{https://github.com/williamdlees/digger/blob/main/tests/human/IGH/IMGT000035/comparison\_results.jpg}{comparison\_results.jpg} summarises functional annotations found by digger and IMGT, according to which,
digger annotated as functional all genes so annotated by IMGT, and annotated an additional 2 V\sphinxhyphen{}genes and one J\sphinxhyphen{}gene as functional.

\noindent\sphinxincludegraphics[width=600\sphinxpxdimen]{{igh_results}.jpg}

\sphinxAtStartPar
\sphinxhref{https://github.com/williamdlees/digger/blob/main/tests/human/IGH/IMGT000035/comparison\_results.txt}{comparison\_results.txt} lists the differences in functional analysis in detail.
\sphinxhref{https://github.com/williamdlees/digger/blob/main/tests/human/IGH/IMGT000035/comparison\_results\_notes.txt}{comparison\_results\_notes.txt} adds some commentary: of the two additional
V\sphinxhyphen{}genes annotated by digger as functional, one has unusual variations in the RSS, causing IMGT to annotate it as ORF. The other is annotated as ORF on the grounds that it has not been
seen rearranged. The additional J\sphinxhyphen{}gene is currently annotated by IMGT in the assembly as ORF, although it is listed in the IMGT gene table as functional.

\sphinxAtStartPar
There is not, at present, a clear set of accepted criteria for categorisation of functionality, and minor differences of this nature are to be expected. Over the next few years, we expect to see comparisons of genomic
sequencing of the loci with the expressed repertoire across multiple subjects, and this should allow a deeper understanding to develop. Overall, the comparison of digger results with the manually supervised curation
at IMGT shows a good level of agreement. It is possible that results may change from those noted here, as they are based on downloaded data which may be revised over time.


\section{References}
\label{\detokenize{examples/human_igh:references}}
\sphinxAtStartPar
Lefranc et al., 2015, IMGT®, the international ImMunoGeneTics information system® 25 years on. \sphinxstyleemphasis{Nucleic Acids Res.} \sphinxhref{https://doi.org/10.1093\%2Fnar\%2Fgku1056}{doi: 10.1093/nar/gku1056}.

\sphinxAtStartPar
Watson et al., 2013, Complete haplotype sequence of the human immunoglobulin heavy\sphinxhyphen{}chain variable, diversity, and joining genes and characterization of allelic and copy\sphinxhyphen{}number variation. \sphinxstyleemphasis{Am J Hum Genet} \sphinxhref{http://dx.doi.org/10.1016/j.ajhg.2013.03.004}{doi: 10.1016/j.ajhg.2013.03.004}

\sphinxAtStartPar
Schneider et al., 2017, Evaluation of GRCh38 and de novo haploid genome assemblies demonstrates the enduring quality of the reference assembly. \sphinxstyleemphasis{Genome Research} \sphinxhref{https://genome.cshlp.org/content/27/5/849}{doi: 10.1101/gr.213611.116}

\sphinxstepscope


\chapter{Annotating the rhesus macaque IGH locus}
\label{\detokenize{examples/rhesus_igh:annotating-the-rhesus-macaque-igh-locus}}\label{\detokenize{examples/rhesus_igh::doc}}
\sphinxAtStartPar
In this example we will see how to:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Create position weight matrices from existing IMGT annotations

\item {} 
\sphinxAtStartPar
Use the underlying commands that digger calls, and understand how they might be useful when annotating multiple contigs or scaffolds

\end{itemize}

\sphinxAtStartPar
IMGT has identified scaffolds in the 2006 rhesus macaque reference assembly, Mmul\_051212, which lie within the IGH locus. Here we will bring them together in a single file and annotate them with motifs derived from the current reference assembly, rhemac10 (Mmul\_10).
While this example is somewhat artificial, in that the scaffolds could equally well be handled individually using the digger command, the approach is useful where the number of sequences to be processed
is large. The example also serves to show how the individual commands in the package can be used. This provides some additional flexibility, for example in tuning the blast searches, and also illustrates how they work together.
The comparison with IMGT’s annotation of Mmul\_051212, and a script to reproduce this example using the steps below, can be found \sphinxhref{https://github.com/williamdlees/digger/tree/main/tests/rhesus\_macaque/IGH/Mmul\_521212}{in digger’s Git repository}.


\section{Data}
\label{\detokenize{examples/rhesus_igh:data}}
\sphinxAtStartPar
As in the previous example, the rhesus IGH germline reference set can be downloaded from IMGT with the \sphinxhref{https://williamdlees.github.io/receptor\_utils/\_build/html/introduction.html}{receptor\_utils} command
\sphinxcode{\sphinxupquote{extract\_refs}} (receptor\_utils is installed as part of digger’s installation). However, the rhesus IG gapped V\sphinxhyphen{}genes provided by IMGT contain additional inserted codons relative to
the conventional IMGT alignment. As digger (alongside other tools) expects the conventional alignment, a further step is needed to realign the gapped sequences, using the receptor\_utils
tool \sphinxcode{\sphinxupquote{fix\_macaque\_gaps}}. The following commands will prepare the reference data:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}} \PYG{n}{extract\PYGZus{}refs} \PYG{o}{\PYGZhy{}}\PYG{n}{L} \PYG{n}{IGH} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Macaca mulatta}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{o}{\PYGZgt{}} \PYG{n}{fix\PYGZus{}macaque\PYGZus{}gaps} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHV\PYGZus{}gapped}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
    \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHV\PYGZus{}gapped\PYGZus{}fixed}\PYG{o}{.}\PYG{n}{fasta} \PYG{n}{IGH}
\PYG{o}{\PYGZgt{}} \PYG{n}{cat} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHV}\PYG{o}{.}\PYG{n}{fasta} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHD}\PYG{o}{.}\PYG{n}{fasta} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHJ}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
    \PYG{o}{\PYGZgt{}} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHVDJ}\PYG{o}{.}\PYG{n}{fasta}
\end{sphinxVerbatim}

\sphinxAtStartPar
The Mmul\_51212 scaffolds can be downloaded from IMGT as follows:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}} \PYG{n}{parse\PYGZus{}imgt\PYGZus{}annotations} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{save\PYGZus{}sequence} \PYG{n}{NW\PYGZus{}001157919}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
   \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{https://www.imgt.org/ligmdb/view.action?format=IMGT\PYGZam{}id=NW\PYGZus{}001157919}\PYG{l+s+s2}{\PYGZdq{}} \PYGZbs{}
       \PYG{n}{NW\PYGZus{}001157919\PYGZus{}genes}\PYG{o}{.}\PYG{n}{csv} \PYG{n}{IGH}
\PYG{o}{\PYGZgt{}} \PYG{n}{parse\PYGZus{}imgt\PYGZus{}annotations} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{save\PYGZus{}sequence} \PYG{n}{NW\PYGZus{}001122023}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
   \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{https://www.imgt.org/ligmdb/view.action?format=IMGT\PYGZam{}id=NW\PYGZus{}001122023}\PYG{l+s+s2}{\PYGZdq{}} \PYGZbs{}
   \PYG{n}{NW\PYGZus{}001122023\PYGZus{}genes}\PYG{o}{.}\PYG{n}{csv} \PYG{n}{IGH}
\PYG{o}{\PYGZgt{}} \PYG{n}{parse\PYGZus{}imgt\PYGZus{}annotations} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{save\PYGZus{}sequence} \PYG{n}{NW\PYGZus{}001122024}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
   \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{https://www.imgt.org/ligmdb/view.action?format=IMGT\PYGZam{}id=NW\PYGZus{}001122024}\PYG{l+s+s2}{\PYGZdq{}} \PYGZbs{}
       \PYG{n}{NW\PYGZus{}001122024\PYGZus{}genes}\PYG{o}{.}\PYG{n}{csv} \PYG{n}{IGH}
\PYG{o}{\PYGZgt{}} \PYG{n}{parse\PYGZus{}imgt\PYGZus{}annotations} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{save\PYGZus{}sequence} \PYG{n}{NW\PYGZus{}001121239}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
   \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{https://www.imgt.org/ligmdb/view.action?format=IMGT\PYGZam{}id=NW\PYGZus{}001121239}\PYG{l+s+s2}{\PYGZdq{}} \PYGZbs{}
       \PYG{n}{NW\PYGZus{}001121239\PYGZus{}genes}\PYG{o}{.}\PYG{n}{csv} \PYG{n}{IGH}
\PYG{o}{\PYGZgt{}} \PYG{n}{parse\PYGZus{}imgt\PYGZus{}annotations} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{save\PYGZus{}sequence} \PYG{n}{NW\PYGZus{}001121240}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
   \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{https://www.imgt.org/ligmdb/view.action?format=IMGT\PYGZam{}id=NW\PYGZus{}001121240}\PYG{l+s+s2}{\PYGZdq{}} \PYGZbs{}
       \PYG{n}{NW\PYGZus{}001121240\PYGZus{}genes}\PYG{o}{.}\PYG{n}{csv} \PYG{n}{IGH}

\PYG{o}{\PYGZgt{}}\PYG{n}{cat} \PYG{n}{NW\PYGZus{}001157919}\PYG{o}{.}\PYG{n}{fasta} \PYG{n}{NW\PYGZus{}001122023}\PYG{o}{.}\PYG{n}{fasta} \PYG{n}{NW\PYGZus{}001122024}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
       \PYG{n}{NW\PYGZus{}001121239}\PYG{o}{.}\PYG{n}{fasta} \PYG{n}{NW\PYGZus{}001121240}\PYG{o}{.}\PYG{n}{fasta} \PYG{o}{\PYGZgt{}} \PYG{n}{Mmul\PYGZus{}051212}\PYG{o}{.}\PYG{n}{fasta}
\end{sphinxVerbatim}


\section{Preparing position\sphinxhyphen{}weighted matrices}
\label{\detokenize{examples/rhesus_igh:preparing-position-weighted-matrices}}
\sphinxAtStartPar
Digger already has PWMs for rhesus IGH, but for the purpose of this example, we will create a set using the features listed in IMGT’s annotation of the rhemac10 IGH locus, which
has the IMGT accession number IMGT000064. The following commands download the annotation, determine the features, and calculate the PWMs from
features of functional annotations:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}} \PYG{n}{mkdir} \PYG{n}{motifs}
\PYG{o}{\PYGZgt{}} \PYG{n}{cd} \PYG{n}{motifs}
\PYG{o}{\PYGZgt{}} \PYG{n}{parse\PYGZus{}imgt\PYGZus{}annotations} \PYGZbs{}
        \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{http://www.imgt.org/ligmdb/view?format=IMGT\PYGZam{}id=IMGT000064}\PYG{l+s+s2}{\PYGZdq{}} \PYGZbs{}
        \PYG{n}{IMGT000064\PYGZus{}genes}\PYG{o}{.}\PYG{n}{csv} \PYG{n}{IGH}
\PYG{o}{\PYGZgt{}} \PYG{n}{calc\PYGZus{}motifs} \PYG{n}{IMGT000064\PYGZus{}genes}\PYG{o}{.}\PYG{n}{csv}
\end{sphinxVerbatim}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{calc\_motifs}} will create 10 motif files in the directory.

\sphinxAtStartPar
The motifs directory may optionally contain a FASTA file \sphinxcode{\sphinxupquote{conserved\_motifs.fasta}} defining strongly\sphinxhyphen{}conserved nucleotides in the RSS and leader. Only those features
with conserved residues need to be listed in the file. The names follow the filenames used for the PWMs.
The following sequences were derived from Figure 3 of Ngoune et al. (2022) and will be used in this example:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}}\PYG{n}{V}\PYG{o}{\PYGZhy{}}\PYG{n}{HEPTAMER}
\PYG{n}{CAC}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{G}
\PYG{o}{\PYGZgt{}}\PYG{n}{V}\PYG{o}{\PYGZhy{}}\PYG{n}{NONAMER}
\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{AACC}
\PYG{o}{\PYGZgt{}}\PYG{l+m+mi}{5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{D\PYGZhy{}HEPTAMER}
\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{GTG}
\PYG{o}{\PYGZgt{}}\PYG{l+m+mi}{5}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{D\PYGZhy{}NONAMER}
\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{T}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}
\PYG{o}{\PYGZgt{}}\PYG{l+m+mi}{3}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{D\PYGZhy{}HEPTAMER}
\PYG{n}{C}\PYG{o}{\PYGZhy{}}\PYG{n}{C}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{G}
\PYG{o}{\PYGZgt{}}\PYG{l+m+mi}{3}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{D\PYGZhy{}NONAMER}
\PYG{o}{\PYGZhy{}}\PYG{n}{C}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{A}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}
\PYG{o}{\PYGZgt{}}\PYG{n}{J}\PYG{o}{\PYGZhy{}}\PYG{n}{HEPTAMER}
\PYG{n}{C}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{TGTG}
\PYG{o}{\PYGZgt{}}\PYG{n}{J}\PYG{o}{\PYGZhy{}}\PYG{n}{NONAMER}
\PYG{o}{\PYGZhy{}}\PYG{n}{GTT}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{TG}\PYG{o}{\PYGZhy{}}
\end{sphinxVerbatim}

\sphinxAtStartPar
Again, this file is provided for download at the location provided near the top of this example.

\sphinxAtStartPar
While the presence or absence of conserved residues can be a useful guide to the likely functionality of a sequence, please bear in mind that it is a guide only:
exceptions can be expected, particularly where the definitions have been built on limited data.


\section{Annotating the Assembly}
\label{\detokenize{examples/rhesus_igh:annotating-the-assembly}}
\sphinxAtStartPar
The digger command is not able to handle a FASTA file containing multiple contigs, so we will call the underlying tools directly. We start by creating the blast databases and querying against the assembly,
using the reference genes determined in the study:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}} \PYG{n}{makeblastdb} \PYG{o}{\PYGZhy{}}\PYG{o+ow}{in} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHV}\PYG{o}{.}\PYG{n}{fasta} \PYG{o}{\PYGZhy{}}\PYG{n}{dbtype} \PYG{n}{nucl}
\PYG{o}{\PYGZgt{}} \PYG{n}{makeblastdb} \PYG{o}{\PYGZhy{}}\PYG{o+ow}{in} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHD}\PYG{o}{.}\PYG{n}{fasta} \PYG{o}{\PYGZhy{}}\PYG{n}{dbtype} \PYG{n}{nucl}
\PYG{o}{\PYGZgt{}} \PYG{n}{makeblastdb} \PYG{o}{\PYGZhy{}}\PYG{o+ow}{in} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHJ}\PYG{o}{.}\PYG{n}{fasta} \PYG{o}{\PYGZhy{}}\PYG{n}{dbtype} \PYG{n}{nucl}

\PYG{o}{\PYGZgt{}} \PYG{n}{blastn} \PYG{o}{\PYGZhy{}}\PYG{n}{db} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHV}\PYG{o}{.}\PYG{n}{fasta} \PYG{o}{\PYGZhy{}}\PYG{n}{query} \PYG{n}{Mmul\PYGZus{}051212}\PYG{o}{.}\PYG{n}{fasta} \PYG{o}{\PYGZhy{}}\PYG{n}{out} \PYG{n}{mmul\PYGZus{}IGHV}\PYG{o}{.}\PYG{n}{out} \PYGZbs{}
   \PYG{o}{\PYGZhy{}}\PYG{n}{outfmt} \PYG{l+m+mi}{7} \PYG{o}{\PYGZhy{}}\PYG{n}{gapopen} \PYG{l+m+mi}{5} \PYG{o}{\PYGZhy{}}\PYG{n}{gapextend} \PYG{l+m+mi}{5} \PYG{o}{\PYGZhy{}}\PYG{n}{penalty} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1} \PYG{o}{\PYGZhy{}}\PYG{n}{word\PYGZus{}size} \PYG{l+m+mi}{11}
\PYG{o}{\PYGZgt{}} \PYG{n}{blastn} \PYG{o}{\PYGZhy{}}\PYG{n}{db} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHD}\PYG{o}{.}\PYG{n}{fasta} \PYG{o}{\PYGZhy{}}\PYG{n}{query} \PYG{n}{Mmul\PYGZus{}051212}\PYG{o}{.}\PYG{n}{fasta} \PYG{o}{\PYGZhy{}}\PYG{n}{out} \PYG{n}{mmul\PYGZus{}IGHD}\PYG{o}{.}\PYG{n}{out} \PYGZbs{}
   \PYG{o}{\PYGZhy{}}\PYG{n}{outfmt} \PYG{l+m+mi}{7} \PYG{o}{\PYGZhy{}}\PYG{n}{gapopen} \PYG{l+m+mi}{5} \PYG{o}{\PYGZhy{}}\PYG{n}{gapextend} \PYG{l+m+mi}{5} \PYG{o}{\PYGZhy{}}\PYG{n}{penalty} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1} \PYG{o}{\PYGZhy{}}\PYG{n}{word\PYGZus{}size} \PYG{l+m+mi}{7} \PYG{o}{\PYGZhy{}}\PYG{n}{evalue} \PYG{l+m+mi}{100}
\PYG{o}{\PYGZgt{}} \PYG{n}{blastn} \PYG{o}{\PYGZhy{}}\PYG{n}{db} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHJ}\PYG{o}{.}\PYG{n}{fasta} \PYG{o}{\PYGZhy{}}\PYG{n}{query} \PYG{n}{Mmul\PYGZus{}051212}\PYG{o}{.}\PYG{n}{fasta} \PYG{o}{\PYGZhy{}}\PYG{n}{out} \PYG{n}{mmul\PYGZus{}IGHJ}\PYG{o}{.}\PYG{n}{out} \PYGZbs{}
   \PYG{o}{\PYGZhy{}}\PYG{n}{outfmt} \PYG{l+m+mi}{7} \PYG{o}{\PYGZhy{}}\PYG{n}{gapopen} \PYG{l+m+mi}{5} \PYG{o}{\PYGZhy{}}\PYG{n}{gapextend} \PYG{l+m+mi}{5} \PYG{o}{\PYGZhy{}}\PYG{n}{penalty} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1} \PYG{o}{\PYGZhy{}}\PYG{n}{word\PYGZus{}size} \PYG{l+m+mi}{7}
\end{sphinxVerbatim}

\sphinxAtStartPar
Note that a higher evalue is used for the D genes, as they can be quite short.

\sphinxAtStartPar
Next we call \sphinxcode{\sphinxupquote{blastresults\_to\_csv}} to convert to a more convenient format:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}} \PYG{n}{blastresults\PYGZus{}to\PYGZus{}csv} \PYG{n}{mmul\PYGZus{}IGHV}\PYG{o}{.}\PYG{n}{out} \PYG{n}{mmul\PYGZus{}ighvdj\PYGZus{}}
\PYG{o}{\PYGZgt{}} \PYG{n}{blastresults\PYGZus{}to\PYGZus{}csv} \PYG{n}{mmul\PYGZus{}IGHD}\PYG{o}{.}\PYG{n}{out} \PYG{n}{mmul\PYGZus{}ighvdj\PYGZus{}} \PYG{o}{\PYGZhy{}}\PYG{n}{a}
\PYG{o}{\PYGZgt{}} \PYG{n}{blastresults\PYGZus{}to\PYGZus{}csv} \PYG{n}{mmul\PYGZus{}IGHJ}\PYG{o}{.}\PYG{n}{out} \PYG{n}{mmul\PYGZus{}ighvdj\PYGZus{}} \PYG{o}{\PYGZhy{}}\PYG{n}{a}
\end{sphinxVerbatim}

\sphinxAtStartPar
The commands instruct the tool to create merged files containing V,D and J hits. This is achieved by specifying the same prefix on each command \sphinxcode{\sphinxupquote{(mmul\_ighvdj\_)}} and using the \sphinxcode{\sphinxupquote{\sphinxhyphen{}a}} (append) option.
The records created by blastn contain the name of the contig in which a hit was found. \sphinxcode{\sphinxupquote{blastresults\_to\_csv}} will create one file per contig. The names contain the ID of the contig in
\sphinxcode{\sphinxupquote{Mmul\_051212.fasta}}, except that they are modified where necessary to ensure file system compatibility.

\sphinxAtStartPar
We now call \sphinxcode{\sphinxupquote{find\_alignments}} to process the annotations:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}} \PYG{n}{find\PYGZus{}alignments} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHVDJ}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
       \PYG{n}{Mmul\PYGZus{}051212}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
       \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{mmul\PYGZus{}ighvdj\PYGZus{}nw\PYGZus{}*.csv}\PYG{l+s+s2}{\PYGZdq{}} \PYGZbs{}
       \PYG{o}{\PYGZhy{}}\PYG{n}{ref} \PYG{n}{imgt}\PYG{p}{,}\PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHVDJ}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
       \PYG{o}{\PYGZhy{}}\PYG{n}{align} \PYG{n}{Macaca\PYGZus{}mulatta\PYGZus{}IGHV\PYGZus{}gapped\PYGZus{}fixed}\PYG{o}{.}\PYG{n}{fasta} \PYGZbs{}
       \PYG{o}{\PYGZhy{}}\PYG{n}{motif\PYGZus{}dir} \PYG{n}{motifs} \PYGZbs{}
       \PYG{n}{Mmul\PYGZus{}051212}\PYG{o}{.}\PYG{n}{csv}
\end{sphinxVerbatim}

\sphinxAtStartPar
Note that the third argument, \sphinxcode{\sphinxupquote{"mmul\_ighvdj\_nw\_*.csv"}}, contains a wildcard that will match all the files produced in the previous step. It is quoted to avoid expansion by the shell.
V\sphinxhyphen{}genes in the annotation will be annotated and gapped using the IMGT set as a template (with fixed gaps).
\sphinxcode{\sphinxupquote{find\_alignments}} will attempt to deduce the sense in which to annotate each segment. This is helpful in this case as the contigs vary in their orientation.  Note that we are
specifying the location of the motifs directory created in the previous step rather than the species and locus, which would cause digger to use the built\sphinxhyphen{}in tables.


\section{Comparing the output to the study’s annotation}
\label{\detokenize{examples/rhesus_igh:comparing-the-output-to-the-study-s-annotation}}
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{compare\_annotations}} is not capable of handling the output from multiple sequences in the same file, so unfortunately we need to split the results up for the comparison:
\begin{quote}

\sphinxAtStartPar
\textgreater{} head \sphinxhyphen{}n 1 Mmul\_051212.csv \textgreater{} mmul\_header.csv

\sphinxAtStartPar
\textgreater{} cp mmul\_header.csv NW\_001157919\_digger.csv
\textgreater{} grep NW\_001157919 Mmul\_051212.csv \textgreater{}\textgreater{} NW\_001157919\_digger.csv

\sphinxAtStartPar
\textgreater{} cp mmul\_header.csv NW\_001122023\_digger.csv
\textgreater{} grep NW\_001122023 Mmul\_051212.csv \textgreater{}\textgreater{} NW\_001122023\_digger.csv

\sphinxAtStartPar
\textgreater{} cp mmul\_header.csv NW\_001122024\_digger.csv
\textgreater{} grep NW\_001122024 Mmul\_051212.csv \textgreater{}\textgreater{} NW\_001122024\_digger.csv

\sphinxAtStartPar
\textgreater{} cp mmul\_header.csv NW\_001121239\_digger.csv
\textgreater{} grep NW\_001121239 Mmul\_051212.csv \textgreater{}\textgreater{} NW\_001121239\_digger.csv

\sphinxAtStartPar
\textgreater{} cp mmul\_header.csv NW\_001121240\_digger.csv
\textgreater{} grep NW\_001121240 Mmul\_051212.csv \textgreater{}\textgreater{} NW\_001121240\_digger.csv

\sphinxAtStartPar
\textgreater{} compare\_annotations NW\_001157919\_digger.csv NW\_001157919\_genes.csv forward NW\_001157919\_comp
\textgreater{} compare\_annotations NW\_001122023\_digger.csv NW\_001122023\_genes.csv forward NW\_001122023\_comp
\textgreater{} compare\_annotations NW\_001122024\_digger.csv NW\_001122024\_genes.csv forward NW\_001122024\_comp
\textgreater{} compare\_annotations NW\_001121239\_digger.csv NW\_001121239\_genes.csv forward NW\_001121239\_comp
\textgreater{} compare\_annotations NW\_001121240\_digger.csv NW\_001121240\_genes.csv forward NW\_001121240\_comp
\end{quote}

\sphinxAtStartPar
Scaffold\sphinxhyphen{}by\sphinxhyphen{}scaffold comparisons are provided in \sphinxhref{https://github.com/williamdlees/digger/tree/main/tests/rhesus\_macaque/IGH/Mmul\_051212}{Github}.
and an overall comparison is provided \sphinxhref{https://github.com/williamdlees/digger/tree/main/tests/rhesus\_macaque/IGH/Mmul\_051212/comparison\_notes.txt}{here}.
One sequence, in NW\_001121240, is annotated as functional by digger but not by IMGT, who report no V\sphinxhyphen{}RS. Digger identifies a different start co\sphinxhyphen{}ordinate for the V\sphinxhyphen{}REGION,
and finds a potentially functional RSS. Two V\sphinxhyphen{}sequences are identified as functional by IMGT but not by digger; one of thes has Ns in the leader, while the other
lies at the extreme 5’ end of the scaffold and the RSS is not fully represented: these issues caused digger not to annotate the sequences as functional.

\sphinxAtStartPar
Digger identified a total of 13 potentially functional D\sphinxhyphen{}genes not annotated by IMGT, across four of the five scaffolds, while IMGT annotated D\sphinxhyphen{}genes only in NW\_001121239. The macaque IGHD genes are known
to occupy a small, distinct, region towards the 3’ end of the IGH locus. It would therefore be reasonable to expect them to be located in a single scaffold, and to be
distinct from the V\sphinxhyphen{}genes. However, given the sequencing technology available for sequencing and assembly when the scaffolds were created, and bearing
in mind the short length of the D\sphinxhyphen{}genes, it is possible that the D\sphinxhyphen{}locus was not correctly assembled. Another reason for suspecting this is that two of the D\sphinxhyphen{}sequences
identified by Digger are extremely short, at 3nt and 1nt, and yet appear to be flanked by functional RSS. In contrast, in an
\sphinxhref{https://github.com/williamdlees/digger/tree/main/tests/rhesus\_macaque/IGH/IMGT000064}{annotation of the rhemac10 assembly}, Digger identified only one D\sphinxhyphen{}gene
not annotated by IMGT (this was also outside the D locus).


\section{References}
\label{\detokenize{examples/rhesus_igh:references}}
\sphinxAtStartPar
Ngoune et al., 2022, IMGT® Biocuration and Analysis of the Rhesus Monkey IG Loci. \sphinxstyleemphasis{Vaccines} \sphinxhref{https://www.mdpi.com/2076-393X/10/3/394\#}{doi: 10.3390/vaccines10030394}.

\sphinxAtStartPar
Warren et al., 2020, Sequence Diversity Analyses of an Improved Rhesus Macaque Genome Enhance Its Biomedical Utility. \sphinxstyleemphasis{Science} \sphinxhref{https://doi.org/10.1126/science.abc6617}{doi: 10.1126/science.abc6617}.

\sphinxAtStartPar
Gibbs et al., 2007, Evolutionary and biomedical insights from the rhesus macaque genome. \sphinxstyleemphasis{Science} \sphinxhref{https://doi.org/10.1126/science.1139247}{doi: 10.1126/science.1139247}.

\sphinxstepscope


\chapter{Additional Examples}
\label{\detokenize{examples/additional_examples:additional-examples}}\label{\detokenize{examples/additional_examples::doc}}
\sphinxAtStartPar
Additional examples, covering the IG loci of human and rhesus macaque, can be found in the \sphinxhref{https://github.com/williamdlees/digger/tree/main/tests}{tests folder} of the digger Github repo.
In each case, the example consists of a script file, \sphinxcode{\sphinxupquote{run\_digger.bat}}, which will download necessary data and conduct the analysis, and the analysis results.
Despite its extension, \sphinxcode{\sphinxupquote{run\_digger.bat}} will run under a Linux shell (\sphinxcode{\sphinxupquote{source run\_digger.bat}}), or under Windows. For Windows, the
\sphinxhref{https://www.gnu.org/software/coreutils/}{Gnu core utilities}, or some other implementation of simple Linux shell commands, are required.

\sphinxstepscope


\chapter{Commandline Usage}
\label{\detokenize{usage:commandline-usage}}\label{\detokenize{usage:usage}}\label{\detokenize{usage::doc}}
\sphinxstepscope


\section{blastresults\_to\_csv}
\label{\detokenize{tools/blastresults_to_csv:blastresults-to-csv}}\label{\detokenize{tools/blastresults_to_csv:id1}}\label{\detokenize{tools/blastresults_to_csv::doc}}
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{blastresults\_to\_csv}} converts the output of a blast search in ‘format 7’ into csv format. If there were multiple query sequences, results are split into separate files.
Please refer to \DUrole{xref,std,std-ref}{rhesus\_igh} for example usage of this and the other ‘individual’ commands.

\sphinxAtStartPar

\sphinxAtStartPar
Convert blast file format 7 to one or more CSVs


\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{usage}\PYG{p}{:} \PYG{n}{blastresults\PYGZus{}to\PYGZus{}csv} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{a}\PYG{p}{]} \PYG{n}{infile} \PYG{n}{out\PYGZus{}prefix}
\end{sphinxVerbatim}


\subsection{Positional Arguments}
\label{\detokenize{tools/blastresults_to_csv:positional-arguments}}\begin{optionlist}{3cm}
\item [infile]  
\sphinxAtStartPar
the blast file
\item [out\_prefix]  
\sphinxAtStartPar
prefix for csv files
\end{optionlist}


\subsection{Named Arguments}
\label{\detokenize{tools/blastresults_to_csv:named-arguments}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}a, \sphinxhyphen{}\sphinxhyphen{}append]  
\sphinxAtStartPar
append to existing output files

\sphinxAtStartPar
Default: False
\end{optionlist}

\sphinxstepscope


\section{calc\_motifs}
\label{\detokenize{tools/calc_motifs:calc-motifs}}\label{\detokenize{tools/calc_motifs:id1}}\label{\detokenize{tools/calc_motifs::doc}}
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{calc\_motifs}} creates motif files for RSS and leader fields, based on the features produced by {\hyperref[\detokenize{tools/parse_imgt_annotations:parse-imgt-annotations}]{\sphinxcrossref{\DUrole{std,std-ref}{parse\_imgt\_annotations}}}}. Only annotations
of sequences marked as ‘functional’ are considered. Please refer to \DUrole{xref,std,std-ref}{rhesus\_igh} for example usage of this and the other ‘individual’ commands.

\sphinxAtStartPar

\sphinxAtStartPar
Given a set of gene features, create motif matrices


\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{usage}\PYG{p}{:} \PYG{n}{calc\PYGZus{}motifs} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{n}{feat\PYGZus{}file}
\end{sphinxVerbatim}


\subsection{Positional Arguments}
\label{\detokenize{tools/calc_motifs:positional-arguments}}\begin{optionlist}{3cm}
\item [feat\_file]  
\sphinxAtStartPar
feature file, created, for example, by parse\_imgt\_annotations
\end{optionlist}

\sphinxstepscope


\section{compare\_annotations}
\label{\detokenize{tools/compare_annotations:compare-annotations}}\label{\detokenize{tools/compare_annotations:id1}}\label{\detokenize{tools/compare_annotations::doc}}
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{compare\_annotations}} compares annotations produced by digger with IMGT’s annotations as summarised by parse\_annotations. Three files are produced:
\sphinxhyphen{} a .jpg showing Venn diagrams of identified functional annotations
\sphinxhyphen{} a .txt file summarising the specific functional sequences that were only identified in one annotation as opposed to both
\sphinxhyphen{} a .csv file listing agreements and differences of all sequences annotated by either mmethod
Please refer to {\hyperref[\detokenize{examples/human_igh:human-igh}]{\sphinxcrossref{\DUrole{std,std-ref}{Annotating the human IGH locus}}}} for example usage.

\sphinxAtStartPar

\sphinxAtStartPar
Compare digger results to an IMGT annotation


\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{usage}\PYG{p}{:} \PYG{n}{compare\PYGZus{}annotations} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{nc}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{filter\PYGZus{}annot} \PYG{n}{FILTER\PYGZus{}ANNOT}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{comp\PYGZus{}name} \PYG{n}{COMP\PYGZus{}NAME}\PYG{p}{]} \PYG{n}{digger\PYGZus{}results} \PYG{n}{annotation\PYGZus{}file} \PYG{n}{sense} \PYG{n}{outfile}
\end{sphinxVerbatim}


\subsection{Positional Arguments}
\label{\detokenize{tools/compare_annotations:positional-arguments}}\begin{optionlist}{3cm}
\item [digger\_results]  
\sphinxAtStartPar
Digger results file
\item [annotation\_file]  
\sphinxAtStartPar
IMGT annotation produced by parse\_imgt\_assembly\_x.py
\item [sense]  
\sphinxAtStartPar
Sense of annotation compared to digger results (forward or reverse)
\item [outfile]  
\sphinxAtStartPar
Output file name (will create .csv, .jpg, .txt
\end{optionlist}


\subsection{Named Arguments}
\label{\detokenize{tools/compare_annotations:named-arguments}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}nc]  
\sphinxAtStartPar
include sequences for leader and rss

\sphinxAtStartPar
Default: False
\item [\sphinxhyphen{}\sphinxhyphen{}filter\_annot]  
\sphinxAtStartPar
filter IMGT annotations by sense (forward or reverse)
\item [\sphinxhyphen{}\sphinxhyphen{}comp\_name]  
\sphinxAtStartPar
name to use for comparison (default IMGT)
\end{optionlist}

\sphinxstepscope


\section{digger}
\label{\detokenize{tools/digger:digger}}\label{\detokenize{tools/digger:id1}}\label{\detokenize{tools/digger::doc}}
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{digger}} annotates a single reference assembly, using BLAST to search the assembly for potential germline sequences. It requires an initial reference set
for BLAST to use: this could come from a similar species, or a former annotation.
Please refer to {\hyperref[\detokenize{examples/human_igh:human-igh}]{\sphinxcrossref{\DUrole{std,std-ref}{Annotating the human IGH locus}}}} for example usage.

\sphinxAtStartPar

\sphinxAtStartPar
Find functional and nonfunctional genes in a single assembly sequence


\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{usage}\PYG{p}{:} \PYG{n}{digger} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{species} \PYG{n}{SPECIES}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{motif\PYGZus{}dir} \PYG{n}{MOTIF\PYGZus{}DIR}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{locus} \PYG{n}{LOCUS}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{v\PYGZus{}ref} \PYG{n}{V\PYGZus{}REF}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{d\PYGZus{}ref} \PYG{n}{D\PYGZus{}REF}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{j\PYGZus{}ref} \PYG{n}{J\PYGZus{}REF}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{v\PYGZus{}ref\PYGZus{}gapped} \PYG{n}{V\PYGZus{}REF\PYGZus{}GAPPED}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{ref\PYGZus{}comp} \PYG{n}{REF\PYGZus{}COMP}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{sense} \PYG{n}{SENSE}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{keepwf}\PYG{p}{]}
              \PYG{n}{assembly\PYGZus{}file} \PYG{n}{output\PYGZus{}file}
\end{sphinxVerbatim}


\subsection{Positional Arguments}
\label{\detokenize{tools/digger:positional-arguments}}\begin{optionlist}{3cm}
\item [assembly\_file]  
\sphinxAtStartPar
assembly sequence to search
\item [output\_file]  
\sphinxAtStartPar
output file (csv)
\end{optionlist}


\subsection{Named Arguments}
\label{\detokenize{tools/digger:named-arguments}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}species]  
\sphinxAtStartPar
use motifs for the specified species provided with the package
\item [\sphinxhyphen{}motif\_dir]  
\sphinxAtStartPar
pathname to directory containing motif probability files
\item [\sphinxhyphen{}locus]  
\sphinxAtStartPar
locus (default is IGH)
\item [\sphinxhyphen{}v\_ref]  
\sphinxAtStartPar
set of V reference genes to use as starting point for search
\item [\sphinxhyphen{}d\_ref]  
\sphinxAtStartPar
set of D reference genes to use as starting point for search
\item [\sphinxhyphen{}j\_ref]  
\sphinxAtStartPar
set of J reference genes to use as starting point for search
\item [\sphinxhyphen{}v\_ref\_gapped]  
\sphinxAtStartPar
IMGT\sphinxhyphen{}gapped v\sphinxhyphen{}reference set used to determine alignment of novel sequences
\item [\sphinxhyphen{}ref\_comp]  
\sphinxAtStartPar
ungapped reference set(s) to compare to: name and reference file separated by comma eg mouse,mouse.fasta (may be repeated multiple times)
\item [\sphinxhyphen{}sense]  
\sphinxAtStartPar
sense in which to read the assembly (forward or reverse) (if omitted will select automatically)
\item [\sphinxhyphen{}keepwf]  
\sphinxAtStartPar
keep working files after processing has completed

\sphinxAtStartPar
Default: False
\end{optionlist}

\sphinxAtStartPar
At least one file containing reference genes must be provided. You can, for example, supply \sphinxcode{\sphinxupquote{v\_ref}}, \sphinxcode{\sphinxupquote{d\_ref}} and \sphinxcode{\sphinxupquote{j\_ref}}, or just \sphinxcode{\sphinxupquote{v\_ref}}. Digger will annotate whatever genes are discovered with the corresponding set(s).
In practice, the sets do not have to be that good a match: BLAST will identify partial matches, and Digger’s logic will extend the match to a full gene, including canonical RSS and leader (using the \sphinxcode{\sphinxupquote{motif}} folder).

\sphinxAtStartPar
Digger requires a set of postion\sphinxhyphen{}weighted matrices, to identify RSS and leader. It is also possible to specify conserved locations of motifs. This \sphinxtitleref{motif} data should be stored in a \sphinxcode{\sphinxupquote{motif}} folder. Motifs for
human and rhesus macaque IG are built in to the package, and may be used with \sphinxcode{\sphinxupquote{\sphinxhyphen{}species}} by specifying either \sphinxcode{\sphinxupquote{human}} or \sphinxcode{\sphinxupquote{rhesus\_macaque}}. The species is used in conjunction with \sphinxcode{\sphinxupquote{\sphinxhyphen{}locus}} to determine
the correct motifs. Alternatively, \sphinxcode{\sphinxupquote{\sphinxhyphen{}motif\_dir}} can be used to specify custom motifs created outside of the package. Please refer to {\hyperref[\detokenize{tools/calc_motifs:calc-motifs}]{\sphinxcrossref{\DUrole{std,std-ref}{calc\_motifs}}}} and to \DUrole{xref,std,std-ref}{rhesus\_igh} for further details
on custom motifs.

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{v\_ref\_gapped}} is used to gep v\sphinxhyphen{}sequences correctly in order to identify conserved codons and so on. Again these sequences do not need to be that good a match in practice. The sequences \sphinxstylestrong{must be IMGT aligned with
no extraneous codons}. Note in particular that IMGT has introduced insertions into macaque alignments in recent years. \sphinxstylestrong{Sets with these insertions should not be used}.

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{ref\_comp}} allows you to specify that you would like annotated sequences to be compared with sequences in a set. You can include as many different sets as you wish. The output file will contain columns
for each of these, listing the closest sequence found and the proximity (\%, and number of nucleotides).

\sphinxAtStartPar
If you choose not to specify the \sphinxcode{\sphinxupquote{sense}}, Digger will select the sense that elicits the highest number of hits and the highest evalue (results are shown in the output so that you can decide whether it has made the right choice,
and whether you wish to annotate in both senses)

\sphinxstepscope


\section{find\_alignments}
\label{\detokenize{tools/find_alignments:find-alignments}}\label{\detokenize{tools/find_alignments:id1}}\label{\detokenize{tools/find_alignments::doc}}
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{find\_alignments}} takes the output of a blast search (as formatted by blastresults\_to\_csv.py), checks each identified location for the presence of a gene, and annotates if found.
Please refer to \DUrole{xref,std,std-ref}{rhesus\_igh} for example usage of this and the other ‘individual’ commands.

\sphinxAtStartPar

\sphinxAtStartPar
Find valid genes in a contig given blast matches


\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{usage}\PYG{p}{:} \PYG{n}{find\PYGZus{}alignments} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{species} \PYG{n}{SPECIES}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{motif\PYGZus{}dir} \PYG{n}{MOTIF\PYGZus{}DIR}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{ref} \PYG{n}{REF}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{align} \PYG{n}{ALIGN}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{locus} \PYG{n}{LOCUS}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{sense} \PYG{n}{SENSE}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{debug}\PYG{p}{]} \PYG{n}{germline\PYGZus{}file} \PYG{n}{assembly\PYGZus{}file} \PYG{n}{blast\PYGZus{}file} \PYG{n}{output\PYGZus{}file}
\end{sphinxVerbatim}


\subsection{Positional Arguments}
\label{\detokenize{tools/find_alignments:positional-arguments}}\begin{optionlist}{3cm}
\item [germline\_file]  
\sphinxAtStartPar
reference set used to produce the blast matches
\item [assembly\_file]  
\sphinxAtStartPar
assembly or contig provided to blast
\item [blast\_file]  
\sphinxAtStartPar
results from blast in the format provided by blastresults\_to\_csv (can contain wildcards if there are multiple files, will be matched by glob)
\item [output\_file]  
\sphinxAtStartPar
output file (csv)
\end{optionlist}


\subsection{Named Arguments}
\label{\detokenize{tools/find_alignments:named-arguments}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}species]  
\sphinxAtStartPar
use motifs for the specified species provided with the package
\item [\sphinxhyphen{}motif\_dir]  
\sphinxAtStartPar
use motif probability files present in the specified directory
\item [\sphinxhyphen{}ref]  
\sphinxAtStartPar
ungapped reference to compare to: name and reference file separated by comma eg mouse,mouse.fasta (may be repeated multiple times)
\item [\sphinxhyphen{}align]  
\sphinxAtStartPar
gapped reference file to use for V gene alignments (should contain V genes only), otherwise de novo alignment will be attempted
\item [\sphinxhyphen{}locus]  
\sphinxAtStartPar
locus (default is IGH)
\item [\sphinxhyphen{}sense]  
\sphinxAtStartPar
sense in which to read the assembly (forward or reverse) (will select automatically)
\item [\sphinxhyphen{}debug]  
\sphinxAtStartPar
produce parsing\_errors file with debug information

\sphinxAtStartPar
Default: False
\end{optionlist}

\sphinxstepscope


\section{parse\_imgt\_annotations}
\label{\detokenize{tools/parse_imgt_annotations:parse-imgt-annotations}}\label{\detokenize{tools/parse_imgt_annotations:id1}}\label{\detokenize{tools/parse_imgt_annotations::doc}}
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{parse\_imgt\_annotations}} downloads an IMGT annotation file or or uses a file already downloaded. It parses the file to provide a list of annotated features.
Optionally it will also store the file downloaded, and create a FASTA file containing the annotated assembly.
Please refer to {\hyperref[\detokenize{examples/human_igh:human-igh}]{\sphinxcrossref{\DUrole{std,std-ref}{Annotating the human IGH locus}}}} for example usage.

\sphinxAtStartPar

\sphinxAtStartPar
Given a set of IMGT annotations, build a CSV file containing gene names and co\sphinxhyphen{}ordinates


\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{usage}\PYG{p}{:} \PYG{n}{parse\PYGZus{}imgt\PYGZus{}annotations} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{save\PYGZus{}download} \PYG{n}{SAVE\PYGZus{}DOWNLOAD}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{save\PYGZus{}sequence} \PYG{n}{SAVE\PYGZus{}SEQUENCE}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{save\PYGZus{}imgt\PYGZus{}annots} \PYG{n}{SAVE\PYGZus{}IMGT\PYGZus{}ANNOTS}\PYG{p}{]} \PYG{n}{imgt\PYGZus{}url} \PYG{n}{outfile} \PYG{n}{locus}
\end{sphinxVerbatim}


\subsection{Positional Arguments}
\label{\detokenize{tools/parse_imgt_annotations:positional-arguments}}\begin{optionlist}{3cm}
\item [imgt\_url]  
\sphinxAtStartPar
URL of IMGT annotation, e.g. \sphinxurl{http://www.imgt.org/ligmdb/view?format=IMGT\&id=IMGT000064}, or name of text file containing its contents
\item [outfile]  
\sphinxAtStartPar
Output file (CSV)
\item [locus]  
\sphinxAtStartPar
one of IGH, IGK, IGL, TRA, TRB, TRD, TRG
\end{optionlist}


\subsection{Named Arguments}
\label{\detokenize{tools/parse_imgt_annotations:named-arguments}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}save\_download]  
\sphinxAtStartPar
Save contents of annotation to specified file
\item [\sphinxhyphen{}\sphinxhyphen{}save\_sequence]  
\sphinxAtStartPar
Save sequence to specified file
\item [\sphinxhyphen{}\sphinxhyphen{}save\_imgt\_annots]  
\sphinxAtStartPar
Save IMGT annotations to specified file
\end{optionlist}

\sphinxstepscope


\chapter{Anotation format}
\label{\detokenize{tools/annotation:anotation-format}}\label{\detokenize{tools/annotation:annotation}}\label{\detokenize{tools/annotation::doc}}
\sphinxAtStartPar
This page describes the annotation file produced by digger / find\_alignments


\section{Columns in the Annotation File}
\label{\detokenize{tools/annotation:columns-in-the-annotation-file}}
\sphinxAtStartPar
In addition to the columns in the first table, the file contains the columns in the second table, prefixed by the reference name, for each reference specified with a \sphinxhyphen{}ref argument.


\begin{savenotes}
\sphinxatlongtablestart
\sphinxthistablewithglobalstyle
\begin{longtable}[c]{|\X{25}{100}|\X{75}{100}|}
\sphinxtoprule
\sphinxstyletheadfamily 
\sphinxAtStartPar
Column Name
&\sphinxstyletheadfamily 
\sphinxAtStartPar
Meaning
\\
\sphinxmidrule
\endfirsthead

\multicolumn{2}{c}{\sphinxnorowcolor
    \makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} \textendash{} continued from previous page}}%
}\\
\sphinxtoprule
\sphinxstyletheadfamily 
\sphinxAtStartPar
Column Name
&\sphinxstyletheadfamily 
\sphinxAtStartPar
Meaning
\\
\sphinxmidrule
\endhead

\sphinxbottomrule
\multicolumn{2}{r}{\sphinxnorowcolor
    \makebox[0pt][r]{\sphinxtablecontinued{continues on next page}}%
}\\
\endfoot

\endlastfoot
\sphinxtableatstartofbodyhook

\sphinxAtStartPar
contig
&
\sphinxAtStartPar
ID of the sequence in which the gene or pseudogene was found
\\
\sphinxhline
\sphinxAtStartPar
start
&
\sphinxAtStartPar
start co\sphinxhyphen{}ord of the coding region
\\
\sphinxhline
\sphinxAtStartPar
end
&
\sphinxAtStartPar
end co\sphinxhyphen{}ord of the coding region
\\
\sphinxhline
\sphinxAtStartPar
start\_rev
&
\sphinxAtStartPar
start co\sphinxhyphen{}ord in the reverse\sphinxhyphen{}primed sequence
\\
\sphinxhline
\sphinxAtStartPar
end\_rev
&
\sphinxAtStartPar
end co\sphinxhyphen{}ord in the reverse\sphinxhyphen{}primed sequence
\\
\sphinxhline
\sphinxAtStartPar
sense
&
\sphinxAtStartPar
sense (relative to the input sequence)
\\
\sphinxhline
\sphinxAtStartPar
gene\_type
&
\sphinxAtStartPar
gene type (e.g. IGHV)
\\
\sphinxhline
\sphinxAtStartPar
likelihood
&
\sphinxAtStartPar
likelihood that the RSS is that of a functional gene (compared to a random sequence)
\\
\sphinxhline
\sphinxAtStartPar
l\_part1
&
\sphinxAtStartPar
leader part 1 equence
\\
\sphinxhline
\sphinxAtStartPar
l\_part2
&
\sphinxAtStartPar
leader part 2 sequence
\\
\sphinxhline
\sphinxAtStartPar
v\_heptamer
&
\sphinxAtStartPar
v\sphinxhyphen{}heptamer sequence
\\
\sphinxhline
\sphinxAtStartPar
v\_nonamer
&
\sphinxAtStartPar
v\sphinxhyphen{}nonamer sequence
\\
\sphinxhline
\sphinxAtStartPar
j\_heptamer
&
\sphinxAtStartPar
j\sphinxhyphen{}heptamer sequence
\\
\sphinxhline
\sphinxAtStartPar
j\_nonamer
&
\sphinxAtStartPar
j\sphinxhyphen{}nonamer sequence
\\
\sphinxhline
\sphinxAtStartPar
j\_frame
&
\sphinxAtStartPar
coding frame of the first nucleotide of the j region (0, 1 or 2)
\\
\sphinxhline
\sphinxAtStartPar
d\_3\_heptamer
&
\sphinxAtStartPar
3\sphinxhyphen{}prime d\sphinxhyphen{}heptamer sequence
\\
\sphinxhline
\sphinxAtStartPar
d\_3\_nonamer
&
\sphinxAtStartPar
3\sphinxhyphen{}prime d\sphinxhyphen{}nonamer sequence
\\
\sphinxhline
\sphinxAtStartPar
d\_5\_heptamer
&
\sphinxAtStartPar
5\sphinxhyphen{}prime d\sphinxhyphen{}heptamer sequence
\\
\sphinxhline
\sphinxAtStartPar
d\_5\_nonamer
&
\sphinxAtStartPar
5\sphinxhyphen{}prime d\sphinxhyphen{}nonamer sequence
\\
\sphinxhline
\sphinxAtStartPar
functional
&
\sphinxAtStartPar
functionality (see below)
\\
\sphinxhline
\sphinxAtStartPar
notes
&
\sphinxAtStartPar
annotation notes
\\
\sphinxhline
\sphinxAtStartPar
aa
&
\sphinxAtStartPar
amino acid translation of the coding region
\\
\sphinxhline
\sphinxAtStartPar
v\sphinxhyphen{}gene\_aligned\_aa
&
\sphinxAtStartPar
IMGT\sphinxhyphen{}gapped amino acid translation of the coding sequence (for V\sphinxhyphen{}genes)
\\
\sphinxhline
\sphinxAtStartPar
seq
&
\sphinxAtStartPar
sequence of the coding region
\\
\sphinxhline
\sphinxAtStartPar
seq\_gapped
&
\sphinxAtStartPar
IMGT\sphinxhyphen{}gapped sequence of the coding region (V\sphinxhyphen{}genes only)
\\
\sphinxhline
\sphinxAtStartPar
5\_rss\_start
&
\sphinxAtStartPar
co\sphinxhyphen{}ordinates of the 5\sphinxhyphen{}prime RSS
\\
\sphinxhline
\sphinxAtStartPar
5\_rss\_start\_rev
&\\
\sphinxhline
\sphinxAtStartPar
5\_rss\_end
&\\
\sphinxhline
\sphinxAtStartPar
5\_rss\_end\_rev
&\\
\sphinxhline
\sphinxAtStartPar
3\_rss\_start
&
\sphinxAtStartPar
co\sphinxhyphen{}ordinates of the 3\sphinxhyphen{}prime RSS
\\
\sphinxhline
\sphinxAtStartPar
3\_rss\_start\_rev
&\\
\sphinxhline
\sphinxAtStartPar
3\_rss\_end
&\\
\sphinxhline
\sphinxAtStartPar
3\_rss\_end\_rev
&\\
\sphinxhline
\sphinxAtStartPar
l\_part1\_start
&
\sphinxAtStartPar
co\sphinxhyphen{}ordinates of the leader part 1
\\
\sphinxhline
\sphinxAtStartPar
l\_part1\_start\_rev
&\\
\sphinxhline
\sphinxAtStartPar
l\_part1\_end
&\\
\sphinxhline
\sphinxAtStartPar
l\_part1\_end\_rev
&\\
\sphinxhline
\sphinxAtStartPar
l\_part2\_start
&
\sphinxAtStartPar
co\sphinxhyphen{}ordinates of the leader part 2
\\
\sphinxhline
\sphinxAtStartPar
l\_part2\_start\_rev
&\\
\sphinxhline
\sphinxAtStartPar
l\_part2\_end
&\\
\sphinxhline
\sphinxAtStartPar
l\_part2\_end\_rev
&\\
\sphinxhline
\sphinxAtStartPar
matches
&
\sphinxAtStartPar
number of matches to this start/end region that were produced in the BLAST analysis
\\
\sphinxhline
\sphinxAtStartPar
blast\_match
&
\sphinxAtStartPar
gene in the reference file with the highest match score in this start/end region
\\
\sphinxhline
\sphinxAtStartPar
blast\_score
&
\sphinxAtStartPar
the highest BLAST match score in this start/end region
\\
\sphinxhline
\sphinxAtStartPar
blast\_nt\_diffs
&
\sphinxAtStartPar
the number of nucleotides differing from the most highly scoring reference sequence in this BLAST match
\\
\sphinxhline
\sphinxAtStartPar
evalue
&
\sphinxAtStartPar
evalue of the most highly scoring BLAST match in this start/end region
\\
\sphinxbottomrule
\end{longtable}
\sphinxtableafterendhook
\sphinxatlongtableend
\end{savenotes}

\sphinxAtStartPar
Columns provided for each \sphinxhyphen{}ref:


\begin{savenotes}\sphinxattablestart
\sphinxthistablewithglobalstyle
\centering
\begin{tabular}[t]{|\X{25}{100}|\X{75}{100}|}
\sphinxtoprule
\sphinxstyletheadfamily 
\sphinxAtStartPar
Column Name
&\sphinxstyletheadfamily 
\sphinxAtStartPar
Meaning
\\
\sphinxmidrule
\sphinxtableatstartofbodyhook
\sphinxAtStartPar
\_match
&
\sphinxAtStartPar
ID of the closest matching reference gene
\\
\sphinxhline
\sphinxAtStartPar
\_score
&
\sphinxAtStartPar
score of the closest match
\\
\sphinxhline
\sphinxAtStartPar
\_nt\_diffs
&
\sphinxAtStartPar
number of nucleotides differing from the closest reference sequence
\\
\sphinxbottomrule
\end{tabular}
\sphinxtableafterendhook\par
\sphinxattableend\end{savenotes}


\section{Functionality}
\label{\detokenize{tools/annotation:functionality}}
\sphinxAtStartPar
Functionality is assigned as follows:

\sphinxAtStartPar
Functional
\begin{itemize}
\item {} 
\sphinxAtStartPar
RSS and leader meet or exceed position\sphinxhyphen{}weighted matrix threshold

\item {} 
\sphinxAtStartPar
Highly\sphinxhyphen{}conserved residues agree with the definition for the locus, if a definition has been specified

\item {} 
\sphinxAtStartPar
If a V\sphinxhyphen{}gene, leader starts with ATG, and spliced leader has no stop codons

\item {} 
\sphinxAtStartPar
If a V\sphinxhyphen{}gene, coding region has no stop codons before the cysteine at IMGT position 104

\item {} 
\sphinxAtStartPar
If a V\sphinxhyphen{}gene, conserved residues are at the expected locations

\item {} 
\sphinxAtStartPar
If a J\sphinxhyphen{}gene, donor splice is as expected and coding region has no stop codons

\end{itemize}

\sphinxAtStartPar
ORF
\begin{itemize}
\item {} 
\sphinxAtStartPar
One or more of the above conditions are not met, but no stop codon has been detected

\item {} 
\sphinxAtStartPar
If a V\sphinxhyphen{}gene, leader starts with ATG

\end{itemize}

\sphinxAtStartPar
Pseudo
\begin{itemize}
\item {} 
\sphinxAtStartPar
Coding region contains stop codon(s)

\item {} 
\sphinxAtStartPar
Leader does not start with ATG

\end{itemize}


\chapter{Indices and tables}
\label{\detokenize{index:indices-and-tables}}\begin{itemize}
\item {} 
\sphinxAtStartPar
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\sphinxAtStartPar
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\sphinxAtStartPar
\DUrole{xref,std,std-ref}{search}

\end{itemize}



\renewcommand{\indexname}{Index}
\printindex
\end{document}