<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Anotation format &mdash; Digger 0.5.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="parse_imgt_annotations" href="parse_imgt_annotations.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Digger
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../news.html">Release Notes</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../examples/human_igh.html">Annotating the human IGH locus</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/rhesus_igh.html">Annotating the rhesus macaque IGH locus</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/additional_examples.html">Additional Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage Documentation</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Commandline Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Anotation format</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#columns-in-the-annotation-file">Columns in the Annotation File</a></li>
<li class="toctree-l2"><a class="reference internal" href="#functionality">Functionality</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Digger</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Anotation format</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tools/annotation.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="anotation-format">
<span id="annotation"></span><h1>Anotation format<a class="headerlink" href="#anotation-format" title="Permalink to this heading"></a></h1>
<p>This page describes the annotation file produced by digger / find_alignments</p>
<section id="columns-in-the-annotation-file">
<h2>Columns in the Annotation File<a class="headerlink" href="#columns-in-the-annotation-file" title="Permalink to this heading"></a></h2>
<p>In addition to the columns in the first table, the file contains the columns in the second table, prefixed by the reference name, for each reference specified with a -ref argument.</p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 75%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Column Name</p></th>
<th class="head"><p>Meaning</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>contig</p></td>
<td><p>ID of the sequence in which the gene or pseudogene was found</p></td>
</tr>
<tr class="row-odd"><td><p>start</p></td>
<td><p>start co-ord of the coding region</p></td>
</tr>
<tr class="row-even"><td><p>end</p></td>
<td><p>end co-ord of the coding region</p></td>
</tr>
<tr class="row-odd"><td><p>start_rev</p></td>
<td><p>start co-ord in the reverse-primed sequence</p></td>
</tr>
<tr class="row-even"><td><p>end_rev</p></td>
<td><p>end co-ord in the reverse-primed sequence</p></td>
</tr>
<tr class="row-odd"><td><p>sense</p></td>
<td><p>sense (relative to the input sequence)</p></td>
</tr>
<tr class="row-even"><td><p>gene_type</p></td>
<td><p>gene type (e.g. IGHV)</p></td>
</tr>
<tr class="row-odd"><td><p>gene_start start co-ord of the entire gene including flanking regions</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>gene_end end co-ord of the entire gene including flanking regions</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>gene_start start co-ord of the entire gene including flanking regions in the reverse-primed sequence</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>gene_end end co-ord of the entire gene including flanking regions in the reverse-primed sequence</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>likelihood</p></td>
<td><p>likelihood that the RSS is that of a functional gene (compared to a random sequence)</p></td>
</tr>
<tr class="row-even"><td><p>l_part1</p></td>
<td><p>leader part 1 equence</p></td>
</tr>
<tr class="row-odd"><td><p>l_part2</p></td>
<td><p>leader part 2 sequence</p></td>
</tr>
<tr class="row-even"><td><p>v_heptamer</p></td>
<td><p>v-heptamer sequence</p></td>
</tr>
<tr class="row-odd"><td><p>v_nonamer</p></td>
<td><p>v-nonamer sequence</p></td>
</tr>
<tr class="row-even"><td><p>j_heptamer</p></td>
<td><p>j-heptamer sequence</p></td>
</tr>
<tr class="row-odd"><td><p>j_nonamer</p></td>
<td><p>j-nonamer sequence</p></td>
</tr>
<tr class="row-even"><td><p>j_frame</p></td>
<td><p>coding frame of the first nucleotide of the j region (0, 1 or 2)</p></td>
</tr>
<tr class="row-odd"><td><p>d_3_heptamer</p></td>
<td><p>3-prime d-heptamer sequence</p></td>
</tr>
<tr class="row-even"><td><p>d_3_nonamer</p></td>
<td><p>3-prime d-nonamer sequence</p></td>
</tr>
<tr class="row-odd"><td><p>d_5_heptamer</p></td>
<td><p>5-prime d-heptamer sequence</p></td>
</tr>
<tr class="row-even"><td><p>d_5_nonamer</p></td>
<td><p>5-prime d-nonamer sequence</p></td>
</tr>
<tr class="row-odd"><td><p>functional</p></td>
<td><p>functionality (see below)</p></td>
</tr>
<tr class="row-even"><td><p>notes</p></td>
<td><p>annotation notes</p></td>
</tr>
<tr class="row-odd"><td><p>aa</p></td>
<td><p>amino acid translation of the coding region</p></td>
</tr>
<tr class="row-even"><td><p>v-gene_aligned_aa</p></td>
<td><p>IMGT-gapped amino acid translation of the coding sequence (for V-genes)</p></td>
</tr>
<tr class="row-odd"><td><p>seq</p></td>
<td><p>sequence of the coding region</p></td>
</tr>
<tr class="row-even"><td><p>seq_gapped</p></td>
<td><p>IMGT-gapped sequence of the coding region (V-genes only)</p></td>
</tr>
<tr class="row-odd"><td><p>5_rss_start</p></td>
<td><p>co-ordinates of the 5-prime RSS</p></td>
</tr>
<tr class="row-even"><td><p>5_rss_start_rev</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>5_rss_end</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>5_rss_end_rev</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>3_rss_start</p></td>
<td><p>co-ordinates of the 3-prime RSS</p></td>
</tr>
<tr class="row-even"><td><p>3_rss_start_rev</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>3_rss_end</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>3_rss_end_rev</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>l_part1_start</p></td>
<td><p>co-ordinates of the leader part 1</p></td>
</tr>
<tr class="row-even"><td><p>l_part1_start_rev</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>l_part1_end</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>l_part1_end_rev</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>l_part2_start</p></td>
<td><p>co-ordinates of the leader part 2</p></td>
</tr>
<tr class="row-even"><td><p>l_part2_start_rev</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>l_part2_end</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>l_part2_end_rev</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>matches</p></td>
<td><p>number of matches to this start/end region that were produced in the BLAST analysis</p></td>
</tr>
<tr class="row-even"><td><p>blast_match</p></td>
<td><p>gene in the reference file with the highest match score in this start/end region</p></td>
</tr>
<tr class="row-odd"><td><p>blast_score</p></td>
<td><p>the highest BLAST match score in this start/end region</p></td>
</tr>
<tr class="row-even"><td><p>blast_nt_diffs</p></td>
<td><p>the number of nucleotides differing from the most highly scoring reference sequence in this BLAST match</p></td>
</tr>
<tr class="row-odd"><td><p>evalue</p></td>
<td><p>evalue of the most highly scoring BLAST match in this start/end region</p></td>
</tr>
</tbody>
</table>
<p>Columns provided for each -ref:</p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 75%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Column Name</p></th>
<th class="head"><p>Meaning</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>_match</p></td>
<td><p>ID of the closest matching reference gene</p></td>
</tr>
<tr class="row-odd"><td><p>_score</p></td>
<td><p>score of the closest match</p></td>
</tr>
<tr class="row-even"><td><p>_nt_diffs</p></td>
<td><p>number of nucleotides differing from the closest reference sequence</p></td>
</tr>
</tbody>
</table>
</section>
<section id="functionality">
<h2>Functionality<a class="headerlink" href="#functionality" title="Permalink to this heading"></a></h2>
<p>Functionality is assigned as follows:</p>
<p>Functional</p>
<ul class="simple">
<li><p>RSS and leader meet or exceed position-weighted matrix threshold</p></li>
<li><p>Highly-conserved nucleotides agree with the definition for the locus, if a definition has been specified</p></li>
<li><p>If a V-gene, leader starts with ATG, and spliced leader has no stop codons</p></li>
<li><p>If a V-gene, coding region has no stop codons before the cysteine at IMGT position 104</p></li>
<li><p>If a V-gene, conserved nucleotides are at the expected locations</p></li>
<li><p>If a J-gene, donor splice is as expected and coding region has no stop codons</p></li>
</ul>
<p>ORF</p>
<ul class="simple">
<li><p>One or more of the above conditions are not met, but no stop codon has been detected</p></li>
<li><p>If a V-gene, leader starts with ATG</p></li>
</ul>
<p>Pseudo</p>
<ul class="simple">
<li><p>Coding region contains stop codon(s)</p></li>
<li><p>Leader does not start with ATG</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="parse_imgt_annotations.html" class="btn btn-neutral float-left" title="parse_imgt_annotations" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, William Lees.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>