<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>dig_sequence &mdash; Digger 0.5.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="find_alignments" href="find_alignments.html" />
    <link rel="prev" title="digger" href="digger.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Digger
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../overview.html#digger"><code class="docutils literal notranslate"><span class="pre">digger</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../overview.html#dig-sequence"><code class="docutils literal notranslate"><span class="pre">dig-sequence</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../news.html">Release Notes</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../examples/human_igh.html">Annotating the human IGH locus</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/rhesus_igh.html">Annotating the rhesus macaque IGH locus</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/targeted_annotation.html">Targeted Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/additional_examples.html">Additional Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage Documentation</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../usage.html">Commandline Usage</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="blastresults_to_csv.html">blastresults_to_csv</a></li>
<li class="toctree-l2"><a class="reference internal" href="calc_motifs.html">calc_motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="compare_annotations.html">compare_annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="digger.html">digger</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">dig_sequence</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Sub-commands">Sub-commands</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#fasta">fasta</a></li>
<li class="toctree-l4"><a class="reference internal" href="#single">single</a></li>
<li class="toctree-l4"><a class="reference internal" href="#multi">multi</a></li>
<li class="toctree-l4"><a class="reference internal" href="#multi_seq">multi_seq</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="find_alignments.html">find_alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="parse_imgt_annotations.html">parse_imgt_annotations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="annotation.html">Anotation format</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Digger</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
          <li class="breadcrumb-item"><a href="../usage.html">Commandline Usage</a></li>
      <li class="breadcrumb-item active">dig_sequence</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tools/dig_sequence.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="dig-sequence">
<span id="id1"></span><h1>dig_sequence<a class="headerlink" href="#dig-sequence" title="Permalink to this heading"></a></h1>
<p><code class="docutils literal notranslate"><span class="pre">dig_sequence</span></code> provides targeted annotation of genomic sequences. The sequences can be as small as a single coding region,
or as large as an entire locus. The tool will search for the sequence that best matches a specified target sequence, and
annotate just that best match</p>
<p>Options are available to annotate a single sequence in a FASTA file, a single genbank ID (the sequence will be fetched from
Genbank), or a list of sequences or Genbank IDs specified in a CSV file. For GenBank requests, an email address must be
provided, as this is a requirement of the GenBank API.</p>
<p><p>Annotate a genomic sequence representing the nominated receptor gene</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">dig_sequence</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">{</span><span class="n">fasta</span><span class="p">,</span><span class="n">single</span><span class="p">,</span><span class="n">multi</span><span class="p">,</span><span class="n">multi_seq</span><span class="p">}</span> <span class="o">...</span>
</pre></div>
</div>
<section id="Sub-commands">
<h2>Sub-commands<a class="headerlink" href="#Sub-commands" title="Permalink to this heading"></a></h2>
<section id="fasta">
<h3>fasta<a class="headerlink" href="#fasta" title="Permalink to this heading"></a></h3>
<p>Annotate a single genomic sequence in a FASTA file</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dig_sequence</span> <span class="n">fasta</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">align</span> <span class="n">ALIGN</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">species</span> <span class="n">SPECIES</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">motif_dir</span> <span class="n">MOTIF_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">out_file</span> <span class="n">OUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">debug</span><span class="p">]</span> <span class="n">target</span> <span class="n">germline_file</span> <span class="n">query_file</span>
</pre></div>
</div>
<section id="positional-arguments">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>target</kbd></dt>
<dd><p>Name of nominated sequence in reference set</p>
</dd>
<dt><kbd>germline_file</kbd></dt>
<dd><p>ungapped reference set containing the nominated sequence (FASTA)</p>
</dd>
<dt><kbd>query_file</kbd></dt>
<dd><p>file containing the sequence to annotate (FASTA)</p>
</dd>
</dl>
</section>
<section id="named-arguments">
<h4>Named Arguments<a class="headerlink" href="#named-arguments" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>-align</kbd></dt>
<dd><p>gapped reference set to use for V gene alignments (required for V gene analysis</p>
</dd>
<dt><kbd>-species</kbd></dt>
<dd><p>use motifs for the specified species provided with the package</p>
</dd>
<dt><kbd>-motif_dir</kbd></dt>
<dd><p>use motif probability files present in the specified directory</p>
</dd>
<dt><kbd>-out_file</kbd></dt>
<dd><p>output file (CSV)</p>
</dd>
<dt><kbd>-debug</kbd></dt>
<dd><p>produce parsing_errors file with debug information</p>
<p>Default: False</p>
</dd>
</dl>
</section>
</section>
<section id="single">
<h3>single<a class="headerlink" href="#single" title="Permalink to this heading"></a></h3>
<p>Annotate a single sequence given its genbank accession number</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dig_sequence</span> <span class="n">single</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">align</span> <span class="n">ALIGN</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">species</span> <span class="n">SPECIES</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">motif_dir</span> <span class="n">MOTIF_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">out_file</span> <span class="n">OUT_FILE</span><span class="p">]</span> <span class="n">target</span> <span class="n">germline_file</span> <span class="n">genbank_acc</span> <span class="n">email_addr</span>
</pre></div>
</div>
<section id="positional-arguments_repeat1">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat1" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>target</kbd></dt>
<dd><p>Name of nominated sequence</p>
</dd>
<dt><kbd>germline_file</kbd></dt>
<dd><p>ungapped reference set containing the nominated sequence (FASTA)</p>
</dd>
<dt><kbd>genbank_acc</kbd></dt>
<dd><p>genbank accession number of the sequence to annotate</p>
</dd>
<dt><kbd>email_addr</kbd></dt>
<dd><p>email address to provide to genbank</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat1">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat1" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>-align</kbd></dt>
<dd><p>gapped reference set to use for V gene alignments (required for V gene analysis</p>
</dd>
<dt><kbd>-species</kbd></dt>
<dd><p>use motifs for the specified species provided with the package</p>
</dd>
<dt><kbd>-motif_dir</kbd></dt>
<dd><p>use motif probability files present in the specified directory</p>
</dd>
<dt><kbd>-out_file</kbd></dt>
<dd><p>output file (CSV)</p>
</dd>
</dl>
</section>
</section>
<section id="multi">
<h3>multi<a class="headerlink" href="#multi" title="Permalink to this heading"></a></h3>
<p>Read allele names and corresponding genbank accession numbers from a CSV file</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dig_sequence</span> <span class="n">multi</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">align</span> <span class="n">ALIGN</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">species</span> <span class="n">SPECIES</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">motif_dir</span> <span class="n">MOTIF_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">out_file</span> <span class="n">OUT_FILE</span><span class="p">]</span> <span class="n">locus</span> <span class="n">germline_file</span> <span class="n">query_file</span> <span class="n">email_addr</span>
</pre></div>
</div>
<section id="positional-arguments_repeat2">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat2" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>locus</kbd></dt>
<dd><p>Locus of nominated sequences</p>
</dd>
<dt><kbd>germline_file</kbd></dt>
<dd><p>ungapped reference set containing the nominated sequence (FASTA)</p>
</dd>
<dt><kbd>query_file</kbd></dt>
<dd><p>File containing list of targets and associated genbank accession numbers (CSV)</p>
</dd>
<dt><kbd>email_addr</kbd></dt>
<dd><p>email address to provide to genbank</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat2">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat2" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>-align</kbd></dt>
<dd><p>gapped reference set to use for V gene alignments (required for V gene analysis</p>
</dd>
<dt><kbd>-species</kbd></dt>
<dd><p>use motifs for the specified species provided with the package</p>
</dd>
<dt><kbd>-motif_dir</kbd></dt>
<dd><p>use motif probability files present in the specified directory</p>
</dd>
<dt><kbd>-out_file</kbd></dt>
<dd><p>output file (CSV)</p>
</dd>
</dl>
</section>
</section>
<section id="multi_seq">
<h3>multi_seq<a class="headerlink" href="#multi_seq" title="Permalink to this heading"></a></h3>
<p>Read allele names and genomic sequences from a CSV file</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dig_sequence</span> <span class="n">multi_seq</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">align</span> <span class="n">ALIGN</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">species</span> <span class="n">SPECIES</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">motif_dir</span> <span class="n">MOTIF_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">out_file</span> <span class="n">OUT_FILE</span><span class="p">]</span> <span class="n">locus</span> <span class="n">germline_file</span> <span class="n">query_file</span>
</pre></div>
</div>
<section id="positional-arguments_repeat3">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat3" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>locus</kbd></dt>
<dd><p>Locus of nominated sequences</p>
</dd>
<dt><kbd>germline_file</kbd></dt>
<dd><p>ungapped reference set containing the nominated sequence (FASTA)</p>
</dd>
<dt><kbd>query_file</kbd></dt>
<dd><p>File containing list of targets and genomic sequences (CSV)</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat3">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat3" title="Permalink to this heading"></a></h4>
<dl class="option-list">
<dt><kbd>-align</kbd></dt>
<dd><p>gapped reference set to use for V gene alignments (required for V gene analysis</p>
</dd>
<dt><kbd>-species</kbd></dt>
<dd><p>use motifs for the specified species provided with the package</p>
</dd>
<dt><kbd>-motif_dir</kbd></dt>
<dd><p>use motif probability files present in the specified directory</p>
</dd>
<dt><kbd>-out_file</kbd></dt>
<dd><p>output file (CSV)</p>
</dd>
</dl>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="digger.html" class="btn btn-neutral float-left" title="digger" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="find_alignments.html" class="btn btn-neutral float-right" title="find_alignments" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, William Lees.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>